{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "311cb61f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gen3_validator\n",
    "# loading gen3 bundled jsonschema into dictionary class\n",
    "dd = gen3_validator.dict.DataDictionary('../examples/schema/json/schema_dev.json')\n",
    "dd.parse_schema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a5c5829",
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting list of nodes\n",
    "dd.get_nodes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "353f0be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Returning the jsonschema for a given node\n",
    "dd.return_schema('lipidomics_file.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72cf0ec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install dictionaryutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed235b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import dictionaryutils\n",
    "import json\n",
    "import yaml\n",
    "dd = dictionaryutils.DataDictionary('/Users/harrijh/projects/gen3schemadev/examples/schema/yaml')\n",
    "dd.load_data(directory='/Users/harrijh/projects/gen3schemadev/examples/schema/yaml')\n",
    "dd_resolved = dd.schema\n",
    "\n",
    "for k, v in dd_resolved.items():\n",
    "    with open(f'/Users/harrijh/projects/gen3schemadev/examples/schema/yaml/resolved/{k}.yaml', 'w') as f:\n",
    "        yaml.safe_dump(v, f)\n",
    "\n",
    "\n",
    "with open(f'/Users/harrijh/projects/gen3schemadev/examples/schema/json/gen3_bundled_schema_resolved.json', 'w') as f:\n",
    "    json.dump(dd_resolved, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68684fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "dd = dictionaryutils.load_schemas_from_file('/Users/harrijh/projects/gen3schemadev/examples/schema/json/schema_dev.json')\n",
    "dd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f882f040",
   "metadata": {},
   "source": [
    "# To validate a gen3schemadev input yaml with the input schema\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "930678da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install check-jsonschema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49b369d2",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "!check-jsonschema --schemafile ../src/gen3schemadev/schema/input_schema.yml ../src/gen3schemadev/schema/input_example.yml --verbose"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4b8c26f",
   "metadata": {},
   "source": [
    "# To validate a single gen3 schema yaml with the metaschema\n",
    "- note, the schema file needs to be resolved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1fa2150",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "!check-jsonschema --schemafile ../src/gen3schemadev/schema/gen3_metaschema.yml ../examples/schema/yaml/resolved/lipidomics_file.yaml --verbose"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a76d32b8",
   "metadata": {},
   "source": [
    "if you give a non-resolved schema, it will not validate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20608414",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "!check-jsonschema --schemafile ../src/gen3schemadev/schema/gen3_metaschema.yml ../examples/schema/yaml/lipidomics_file.yaml --verbose"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee115a9e",
   "metadata": {},
   "source": [
    "# Compiling gen3schemadev input yaml to gen3 yamls\n",
    "## Steps:\n",
    "1. Validate input yaml against the input schema\n",
    "2. Load the input yaml into python dictionary\n",
    "3. The compiler should start with an empty data structure, with default values for a gen3 schema. This yaml is defined in the `src/gen3schemadev/schema/gen3_schema_template.yml`\n",
    "4. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe1dda5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.schema.gen3_template import *\n",
    "from gen3schemadev.utils import *\n",
    "from gen3schemadev.schema.input_schema import DataModel\n",
    "import logging\n",
    "\n",
    "# Set up basic logging configuration\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    format=\"%(asctime)s [%(levelname)s] %(message)s\"\n",
    ")\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "metaschema_path = \"../src/gen3schemadev/schema/gen3_metaschema.yml\"\n",
    "converter_template = generate_gen3_template(metaschema_path)\n",
    "metaschema = load_yaml(metaschema_path)\n",
    "\n",
    "# loading input example\n",
    "data = load_yaml('../tests/input_example.yml')\n",
    "validated_model = DataModel.model_validate(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb5d4fa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "converter_template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "286a9960",
   "metadata": {},
   "outputs": [],
   "source": [
    "metaschema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa9dfccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "validated_model.model_dump()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eee5930d",
   "metadata": {},
   "source": [
    "Now we need to read to:\n",
    "1. validate the input yaml against the input metaschema\n",
    "2. now we know the input yaml is validated, we can extract the data from the yaml into a data class which has the structure `input.node.properties.links`\n",
    "3. We then use the data class to populated the converter template for each node\n",
    "4. The populated templates are then written to the output directory\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "234d581f",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96f62327",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dataclasses import dataclass\n",
    "class node:\n",
    "    name: str\n",
    "    description: str\n",
    "    category: str\n",
    "    properties: list\n",
    "    links: list\n",
    "\n",
    "from typing import Any\n",
    "\n",
    "def get_node_data(node: str, data: Any) -> node:\n",
    "    try:\n",
    "        for ent in data.nodes:\n",
    "            if ent.name == node:\n",
    "                return ent\n",
    "        raise ValueError(f\"node '{node}' not found in data.nodes\")\n",
    "    except AttributeError as e:\n",
    "        raise AttributeError(f\"Invalid data structure: {e}\")\n",
    "    except Exception as e:\n",
    "        raise Exception(f\"An error occurred while retrieving node data: {e}\")\n",
    "\n",
    "def get_node_links(node: str, data: Any) -> list[dict]:\n",
    "    links = data.links\n",
    "    node_links = []\n",
    "    for link in links:\n",
    "        if link.child == node:\n",
    "            node_links.append(link.model_dump())\n",
    "    return node_links\n",
    "\n",
    "\n",
    "\n",
    "from dataclasses import dataclass, asdict\n",
    "from typing import Optional, List, Union\n",
    "\n",
    "@dataclass\n",
    "class LinkObj:\n",
    "    name: str\n",
    "    backref: str\n",
    "    label: Optional[str]\n",
    "    target_type: str\n",
    "    multiplicity: str\n",
    "    required: bool\n",
    "\n",
    "    def to_dict(self):\n",
    "        return asdict(self)\n",
    "\n",
    "@dataclass\n",
    "class LinkGroup:\n",
    "    exclusive: bool\n",
    "    required: bool\n",
    "    subgroup: List[dict]\n",
    "\n",
    "    def to_dict(self):\n",
    "        return asdict(self)\n",
    "    \n",
    "def create_core_metadata_link(child_name: str) -> dict:\n",
    "    link_obj = LinkObj(\n",
    "        name=f\"core_metadata_collections\",\n",
    "        backref=f\"{child_name}s\",\n",
    "        label=None,\n",
    "        target_type=\"core_metadata_collection\",\n",
    "        multiplicity=\"one_to_one\",\n",
    "        required=True \n",
    "    )\n",
    "    return link_obj.to_dict()\n",
    "\n",
    "def convert_node_links(links: dict, node_file: bool = False) -> dict:\n",
    "    link_list = []\n",
    "    for link in links:\n",
    "        link_obj = LinkObj(\n",
    "            name=f\"{link['parent']}s\",\n",
    "            backref=f\"{link['child']}s\",\n",
    "            label=None,\n",
    "            target_type=link['parent'],\n",
    "            multiplicity=link['multiplicity'],\n",
    "            required=True  # TODO remove this hard code later, should pull from input yaml\n",
    "        )\n",
    "        link_list.append(link_obj.to_dict())\n",
    "\n",
    "    if node_file:\n",
    "        core_link = create_core_metadata_link(links[0]['child'])\n",
    "        link_list.append(core_link)\n",
    "\n",
    "    if len(link_list) > 1:\n",
    "        group = LinkGroup(\n",
    "            exclusive=False,\n",
    "            required=True,\n",
    "            subgroup=link_list\n",
    "        )\n",
    "        output = group.to_dict()\n",
    "    else:\n",
    "        output = link_list\n",
    "    return output\n",
    "\n",
    "### Note to self, you will now need to find a way to add in the _definitions and _terms references.\n",
    "### You also need to define a standard for the _definitions and _terms references files to read from as templates\n",
    "\n",
    "def get_properties(node_name: str, data: Any) -> list[dict]:\n",
    "    output = []\n",
    "    ent = get_node_data(node_name, data)\n",
    "    props = ent.properties\n",
    "    if props:\n",
    "        for prop in props:\n",
    "            pdict = {\n",
    "                prop.name: {k: v for k, v in prop.model_dump().items() if k != \"name\"}\n",
    "            }\n",
    "            output.append(pdict)\n",
    "    else:\n",
    "        raise Exception(f'No properties found for node {node_name}')\n",
    "    return output\n",
    "\n",
    "\n",
    "def get_category(node_name: str, data: Any) -> str:\n",
    "    ent = get_node_data(node_name, data)\n",
    "    category = ent.category\n",
    "    # If it's an Enum, get its value; otherwise, return as is\n",
    "    if hasattr(category, \"value\"):\n",
    "        return category.value\n",
    "    return category\n",
    "\n",
    "def get_node_value(node_name: str, key: str, data: Any):\n",
    "    \"\"\"\n",
    "    Returns the value of a single key within an node object.\n",
    "\n",
    "    Args:\n",
    "        node_name (str): The name of the node to retrieve.\n",
    "        key (str): The key whose value is to be returned.\n",
    "        data (Any): The data structure containing nodes.\n",
    "\n",
    "    Returns:\n",
    "        The value associated with the specified key in the node object.\n",
    "    \"\"\"\n",
    "    ent = get_node_data(node_name, data)\n",
    "    return ent.model_dump()[key]\n",
    "\n",
    "\n",
    "def populate_template(node_name: str, input_data, template) -> dict:\n",
    "    \"\"\"\n",
    "    Populate a Gen3 schema template dictionary with values from a Pydantic data model.\n",
    "\n",
    "    This function takes an node name, a Pydantic model instance containing node data,\n",
    "    and a Gen3 schema template dictionary. It fills a copy of the template with values\n",
    "    from the input data, applying special logic for certain keys (e.g., 'name', 'category',\n",
    "    'properties', 'links'). If a key from the input data is not found in the template,\n",
    "    it is added with a value of None and a warning is logged.\n",
    "\n",
    "    Args:\n",
    "        node_name (str): The name of the node to populate in the template.\n",
    "        input_data: A Pydantic model instance containing the node's data.\n",
    "        template (dict): A Gen3 schema template dictionary to be populated.\n",
    "\n",
    "    Returns:\n",
    "        dict: A new Gen3 schema template dictionary populated with values from the input data.\n",
    "\n",
    "    Side Effects:\n",
    "        Logs a warning if a key from the input data is not found in the template.\n",
    "    \"\"\"\n",
    "    # ... function body ...\n",
    "    ent = get_node_data(node_name, input_data)\n",
    "    ent_dict = ent.model_dump()\n",
    "    output_schema = template.copy()\n",
    "    \n",
    "    # Checking if node is file category\n",
    "    file_cat = False\n",
    "    if get_node_value(node_name, 'category', input_data) == 'file':\n",
    "        file_cat = True\n",
    "    \n",
    "    for key, value in ent_dict.items():\n",
    "        if key == 'name':\n",
    "            output_schema['id'] = value\n",
    "        elif key == 'category':\n",
    "            output_schema[key] = get_category(node_name, input_data)\n",
    "        elif key == 'properties':\n",
    "            output_schema[key] = get_properties(node_name, input_data)\n",
    "        elif key == 'links':\n",
    "            links = get_node_links(node_name, input_data)\n",
    "            output_schema[key] = convert_node_links(links, node_file=file_cat)\n",
    "        elif key in output_schema:\n",
    "            output_schema[key] = value\n",
    "        else:\n",
    "            logger.warning(f\"Key '{key}' not found in template\")\n",
    "    return output_schema\n",
    "\n",
    "validated_model = DataModel.model_validate(data)\n",
    "# links = get_node_links('lipidomics_file', validated_model)\n",
    "# convert_node_links(links, node_file=True)\n",
    "# get_properties('sample', validated_model)\n",
    "# get_category('sample', validated_model)\n",
    "\n",
    "out_template = populate_template('lipidomics_file', validated_model, converter_template)\n",
    "out_template\n",
    "# write_yaml(out_template, 'output.yml')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb3abe1d",
   "metadata": {},
   "source": [
    "# Testing converter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a14aa6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.schema.gen3_template import *\n",
    "from gen3schemadev.utils import *\n",
    "from gen3schemadev.schema.input_schema import DataModel\n",
    "from gen3schemadev.converter import *\n",
    "\n",
    "\n",
    "\n",
    "# Loading template and metaschema\n",
    "metaschema = get_metaschema()\n",
    "converter_template = generate_gen3_template(metaschema)\n",
    "\n",
    "# loading input example\n",
    "data = load_yaml('../tests/input_example.yml')\n",
    "validated_model = DataModel.model_validate(data)\n",
    "\n",
    "\n",
    "import json\n",
    "print(json.dumps(construct_props('sample', validated_model), indent=4))\n",
    "\n",
    "# construct_props('sample', validated_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03ae89d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "prop = {'sample_tube_type': {'type': 'enum',\n",
    "   'description': 'Sample tube type (enum)',\n",
    "   'required': False,\n",
    "   'enums': [{'name': 'EDTA'}, {'name': 'Heparin'}, {'name': 'Citrate'}]}}\n",
    "first_key = next(iter(prop))\n",
    "first_value = prop[first_key]\n",
    "first_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d93c476b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.schema.gen3_template import *\n",
    "from gen3schemadev.utils import *\n",
    "from gen3schemadev.schema.input_schema import DataModel\n",
    "from gen3schemadev.converter import *\n",
    "\n",
    "data = load_yaml('../tests/input_example.yml')\n",
    "validated_model = DataModel.model_validate(data)\n",
    "\n",
    "from typing import Dict, Any, List\n",
    "\n",
    "def strip_required_field(props_list: list[dict]) -> list[dict]:\n",
    "    \"\"\"\n",
    "    Remove the 'required' field from all property dicts in the input list.\n",
    "    Can use the output of get_properties() for this function\n",
    "\n",
    "    Args:\n",
    "        props_list (list): A list of property dictionaries, where each dictionary has a single key\n",
    "            (the property name) and its value is a dictionary describing the property. For example:\n",
    "                [\n",
    "                    {\n",
    "                        \"project_id\": {\n",
    "                            \"type\": \"string\",\n",
    "                            \"description\": \"Synthetic_Dataset_1\",\n",
    "                            \"required\": True,\n",
    "                            \"enums\": None\n",
    "                        }\n",
    "                    },\n",
    "                    ...\n",
    "                ]\n",
    "\n",
    "    Returns:\n",
    "        list: A new list with the same structure as props_list, but with the 'required'\n",
    "            field removed from each property's dictionary (if present).\n",
    "\n",
    "    Note:\n",
    "        This function expects a list of property definitions as typically returned by\n",
    "        get_properties() in the Gen3 schema conversion workflow.\n",
    "        If you are working with a DataSourceProtocol object, you should first extract the\n",
    "        properties list using the appropriate function.\n",
    "    \"\"\"\n",
    "    new_list = []\n",
    "    for prop in props_list:\n",
    "        if isinstance(prop, dict):\n",
    "            # Each prop is {property_name: property_dict}\n",
    "            new_prop = {}\n",
    "            for k, v in prop.items():\n",
    "                if isinstance(v, dict):\n",
    "                    v = {key: val for key, val in v.items() if key != 'required'}\n",
    "                new_prop[k] = v\n",
    "            new_list.append(new_prop)\n",
    "        else:\n",
    "            new_list.append(prop)\n",
    "    return new_list\n",
    "\n",
    "def get_required_prop_names(props_list: list[dict]) -> List[str]:\n",
    "    \"\"\"\n",
    "    Given a list of property dicts (as from get_properties), return a list of property names\n",
    "    where the property dict has 'required': True.\n",
    "\n",
    "    Args:\n",
    "        props_list (list): List of property dictionaries.\n",
    "\n",
    "    Returns:\n",
    "        List[str]: List of property names with required True.\n",
    "    \"\"\"\n",
    "    required_names = []\n",
    "    for prop in props_list:\n",
    "        if isinstance(prop, dict):\n",
    "            for k, v in prop.items():\n",
    "                if isinstance(v, dict) and v.get(\"required\") is True:\n",
    "                    required_names.append(k)\n",
    "    return required_names\n",
    "\n",
    "project_props = get_properties('sample', validated_model)\n",
    "stripped_props = strip_required_field(project_props)\n",
    "required_names = get_required_prop_names(project_props)\n",
    "stripped_props, required_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca2c83ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_properties('lipidomics_file', validated_model)\n",
    "construct_props('lipidomics_file', validated_model)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f4b42cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_node_data('lipidomics_file', validated_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19335da3",
   "metadata": {},
   "outputs": [],
   "source": [
    "links = get_node_links('lipidomics_file', validated_model)\n",
    "convert_node_links(links, node_file=True)\n",
    "get_node_value('lipidomics_file', 'category', validated_model) == 'data_file'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1226c5de",
   "metadata": {},
   "source": [
    "***\n",
    "## Trying to create template from metaschema\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dc80133",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.schema.gen3_template import *\n",
    "\n",
    "out_template = generate_gen3_template('../src/gen3schemadev/schema/gen3_metaschema.yml')\n",
    "\n",
    "write_yaml(out_template, 'output.yml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b1325f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.schema.gen3_template import *\n",
    "generate_def_template()\n",
    "# generate_setting_template()\n",
    "# generate_terms_template()\n",
    "# generate_core_metadata_template()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bfb81d1",
   "metadata": {},
   "source": [
    "# Testing yaml to bundle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c87fd68b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.utils import bundle_yamls\n",
    "\n",
    "bundle_yamls('../examples/schema/yaml')['subject']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ced05d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "path = \"../examples/schema/yaml/acknowledgement.yaml\"\n",
    "\n",
    "os.path.dirname(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f9d9c05",
   "metadata": {},
   "source": [
    "# Working on validation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97d308cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "93e41fa4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-07 16:44:37,281 [INFO] Initializing DataDictionary with schema path: ../examples//schema/json/schema_dev.json\n",
      "2025-10-07 16:44:37,282 [INFO] Initializing ResolveSchema with schema path: ../examples//schema/json/schema_dev.json\n",
      "2025-10-07 16:44:37,283 [INFO] Starting schema resolution process.\n",
      "2025-10-07 16:44:37,283 [INFO] Reading JSON file from path: ../examples//schema/json/schema_dev.json\n",
      "2025-10-07 16:44:37,285 [INFO] Successfully read JSON schema.\n",
      "2025-10-07 16:44:37,285 [INFO] Retrieving node names from schema.\n",
      "2025-10-07 16:44:37,286 [INFO] Splitting schema into individual node schemas.\n",
      "2025-10-07 16:44:37,286 [INFO] Split schema into individual node schemas.\n",
      "2025-10-07 16:44:37,286 [INFO] Retrieving schema for schema ID: _definitions.yaml\n",
      "2025-10-07 16:44:37,286 [INFO] Retrieved definitions schema.\n",
      "2025-10-07 16:44:37,286 [INFO] Retrieving schema for schema ID: _terms.yaml\n",
      "2025-10-07 16:44:37,286 [INFO] Retrieved terms schema.\n",
      "2025-10-07 16:44:37,287 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,287 [INFO] Resolved references in definitions schema.\n",
      "2025-10-07 16:44:37,287 [INFO] Resolving all references in schema list.\n",
      "2025-10-07 16:44:37,287 [INFO] === Resolving Schema References ===\n",
      "2025-10-07 16:44:37,288 [INFO] Retrieving node names from schema.\n",
      "2025-10-07 16:44:37,288 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,288 [INFO] Resolved demographic.yaml\n",
      "2025-10-07 16:44:37,288 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,288 [INFO] Resolved project.yaml\n",
      "2025-10-07 16:44:37,289 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,289 [INFO] Resolved acknowledgement.yaml\n",
      "2025-10-07 16:44:37,289 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,289 [INFO] Resolved medical_history.yaml\n",
      "2025-10-07 16:44:37,289 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,289 [INFO] Resolved _settings.yaml\n",
      "2025-10-07 16:44:37,290 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,290 [INFO] Resolved program.yaml\n",
      "2025-10-07 16:44:37,290 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,290 [INFO] Resolved sample.yaml\n",
      "2025-10-07 16:44:37,290 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,291 [INFO] Resolved lipidomics_file.yaml\n",
      "2025-10-07 16:44:37,291 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,291 [INFO] Resolved publication.yaml\n",
      "2025-10-07 16:44:37,291 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,292 [INFO] Resolved subject.yaml\n",
      "2025-10-07 16:44:37,292 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:37,292 [INFO] Resolved core_metadata_collection.yaml\n",
      "2025-10-07 16:44:37,293 [INFO] Resolved all references in schema list.\n",
      "2025-10-07 16:44:37,293 [INFO] Converting schema list to JSON format.\n",
      "2025-10-07 16:44:37,293 [INFO] Converted resolved schema list to JSON format.\n",
      "2025-10-07 16:44:37,333 [INFO] Validating 'demographic' with metaschema\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key: demographic.yaml\n",
      "key: project.yaml\n",
      "key: acknowledgement.yaml\n",
      "key: medical_history.yaml\n",
      "key: program.yaml\n",
      "key: sample.yaml\n",
      "key: lipidomics_file.yaml\n",
      "key: publication.yaml\n",
      "key: subject.yaml\n",
      "key: core_metadata_collection.yaml\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-07 16:44:37,528 [INFO] Validating 'project' with metaschema\n",
      "2025-10-07 16:44:37,697 [INFO] Validating 'acknowledgement' with metaschema\n",
      "2025-10-07 16:44:37,865 [INFO] Validating 'medical_history' with metaschema\n",
      "2025-10-07 16:44:38,024 [INFO] Validating 'program' with metaschema\n",
      "2025-10-07 16:44:38,190 [INFO] Validating 'sample' with metaschema\n",
      "2025-10-07 16:44:38,348 [INFO] Validating 'lipidomics_file' with metaschema\n",
      "2025-10-07 16:44:38,507 [INFO] Validating 'publication' with metaschema\n",
      "2025-10-07 16:44:38,666 [INFO] Validating 'subject' with metaschema\n",
      "2025-10-07 16:44:38,830 [INFO] Validating 'core_metadata_collection' with metaschema\n"
     ]
    }
   ],
   "source": [
    "from gen3schemadev.utils import *\n",
    "from gen3schemadev.schema.gen3_template import get_metaschema\n",
    "import subprocess\n",
    "import tempfile\n",
    "from gen3_validator.resolve_schema import ResolveSchema\n",
    "\n",
    "def bundled_schema_to_dict_list(file: str, return_aux: bool = False):\n",
    "    \"\"\"\n",
    "    Reads a bundled Gen3 JSON schema file and returns a list of schema dictionaries.\n",
    "\n",
    "    Args:\n",
    "        file (str): Path to the bundled JSON file containing multiple schemas.\n",
    "        return_aux (bool): If True, return only the auxiliary schemas (definitions/settings/terms).\n",
    "                           If False, return only the main node schemas.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of schema dictionaries extracted from the bundled file.\n",
    "    \"\"\"\n",
    "    resolver = ResolveSchema(file)\n",
    "    resolver.resolve_schema()\n",
    "    bundled = resolver.schema_resolved  # This is a dict-like object\n",
    "    schema_list = []\n",
    "    aux_list = []\n",
    "    aux_schema_names = ['_definitions.yaml', '_settings.yaml', '_terms.yaml']\n",
    "    for k, v in bundled.items():\n",
    "        print(f\"key: {k}\")\n",
    "        if k in aux_schema_names:\n",
    "            print(f\"found auxiliary schema: {k}\")\n",
    "            aux_list.append(v)\n",
    "        else:\n",
    "            schema_list.append(v)\n",
    "\n",
    "    if return_aux:\n",
    "        return aux_list\n",
    "    else:\n",
    "        return schema_list\n",
    "\n",
    "\n",
    "import jsonschema\n",
    "\n",
    "def validate_schema_with_metaschema(schema, metaschema=None, verbose=False):\n",
    "    \"\"\"\n",
    "    Validate a JSON Schema against a metaschema using the check-jsonschema CLI tool.\n",
    "\n",
    "    This function writes the provided schema and metaschema to temporary files and\n",
    "    invokes the external `check-jsonschema` command-line tool to perform validation.\n",
    "\n",
    "    Args:\n",
    "        schema (dict): The JSON Schema to validate.\n",
    "        metaschema (dict, optional): The metaschema to validate against.\n",
    "            If None, you must provide a metaschema explicitly.\n",
    "\n",
    "    Raises:\n",
    "        subprocess.CalledProcessError: If the check-jsonschema command fails.\n",
    "\n",
    "    Returns:\n",
    "        None. Raises an error if validation fails, otherwise completes silently.\n",
    "\n",
    "    Note:\n",
    "        This function does not return a value. It will raise an error if validation fails.\n",
    "        It is intended for use in environments where the check-jsonschema CLI is available.\n",
    "    \"\"\"\n",
    "    logger.info(f\"Validating '{schema.get('id', '')}' with metaschema\")\n",
    "    # Create temp files for schema and metaschema\n",
    "    with tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=False) as schema_file, \\\n",
    "         tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=False) as metaschema_file:\n",
    "        json.dump(schema, schema_file)\n",
    "        schema_file.flush()\n",
    "        json.dump(metaschema, metaschema_file)\n",
    "        metaschema_file.flush()\n",
    "        schema_path = schema_file.name\n",
    "        metaschema_path = metaschema_file.name\n",
    "    \n",
    "    if verbose:\n",
    "        cmd = [\n",
    "            \"check-jsonschema\", \"--verbose\",\n",
    "            \"--schemafile\", metaschema_path,\n",
    "            schema_path\n",
    "        ]\n",
    "    else:\n",
    "        cmd = [\n",
    "            \"check-jsonschema\",\n",
    "            \"--schemafile\", metaschema_path,\n",
    "            schema_path\n",
    "        ]\n",
    "    # Capture output and wait for the subprocess to complete before returning\n",
    "    completed_process = subprocess.run(cmd, capture_output=True, text=True)\n",
    "    if completed_process.returncode != 0:\n",
    "        logger.error(f\"check-jsonschema failed with exit code {completed_process.returncode}\")\n",
    "        # Log stdout/stderr if available\n",
    "        if completed_process.stdout:\n",
    "            logger.error(f\"STDOUT: {completed_process.stdout}\")\n",
    "        if completed_process.stderr:\n",
    "            logger.error(f\"STDERR: {completed_process.stderr}\")\n",
    "        # Do not raise an error, just return\n",
    "        return\n",
    "\n",
    "\n",
    "\n",
    "schema_list = bundled_schema_to_dict_list('../examples//schema/json/schema_dev.json')\n",
    "# schema_list = bundled_schema_to_dict_list('../output/test_schema_bundle.json')\n",
    "metaschema = get_metaschema()\n",
    "\n",
    "for schema in schema_list:\n",
    "    validate_schema_with_metaschema(schema, metaschema=metaschema, verbose=True)\n",
    "\n",
    "# validate_schema_with_metaschema(schema_list[5], metaschema=metaschema, verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6fb38baa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-07 16:44:41,133 [INFO] Initializing DataDictionary with schema path: ../examples//schema/json/schema_dev.json\n",
      "2025-10-07 16:44:41,134 [INFO] Initializing ResolveSchema with schema path: ../examples//schema/json/schema_dev.json\n",
      "2025-10-07 16:44:41,134 [INFO] Starting schema resolution process.\n",
      "2025-10-07 16:44:41,134 [INFO] Reading JSON file from path: ../examples//schema/json/schema_dev.json\n",
      "2025-10-07 16:44:41,136 [INFO] Successfully read JSON schema.\n",
      "2025-10-07 16:44:41,136 [INFO] Retrieving node names from schema.\n",
      "2025-10-07 16:44:41,137 [INFO] Splitting schema into individual node schemas.\n",
      "2025-10-07 16:44:41,137 [INFO] Split schema into individual node schemas.\n",
      "2025-10-07 16:44:41,137 [INFO] Retrieving schema for schema ID: _definitions.yaml\n",
      "2025-10-07 16:44:41,137 [INFO] Retrieved definitions schema.\n",
      "2025-10-07 16:44:41,137 [INFO] Retrieving schema for schema ID: _terms.yaml\n",
      "2025-10-07 16:44:41,138 [INFO] Retrieved terms schema.\n",
      "2025-10-07 16:44:41,138 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,138 [INFO] Resolved references in definitions schema.\n",
      "2025-10-07 16:44:41,139 [INFO] Resolving all references in schema list.\n",
      "2025-10-07 16:44:41,139 [INFO] === Resolving Schema References ===\n",
      "2025-10-07 16:44:41,139 [INFO] Retrieving node names from schema.\n",
      "2025-10-07 16:44:41,140 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,140 [INFO] Resolved demographic.yaml\n",
      "2025-10-07 16:44:41,140 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,141 [INFO] Resolved project.yaml\n",
      "2025-10-07 16:44:41,141 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,141 [INFO] Resolved acknowledgement.yaml\n",
      "2025-10-07 16:44:41,141 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,142 [INFO] Resolved medical_history.yaml\n",
      "2025-10-07 16:44:41,142 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,142 [INFO] Resolved _settings.yaml\n",
      "2025-10-07 16:44:41,142 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,143 [INFO] Resolved program.yaml\n",
      "2025-10-07 16:44:41,143 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,144 [INFO] Resolved sample.yaml\n",
      "2025-10-07 16:44:41,144 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,144 [INFO] Resolved lipidomics_file.yaml\n",
      "2025-10-07 16:44:41,145 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,145 [INFO] Resolved publication.yaml\n",
      "2025-10-07 16:44:41,145 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,145 [INFO] Resolved subject.yaml\n",
      "2025-10-07 16:44:41,146 [INFO] Resolving references in schema.\n",
      "2025-10-07 16:44:41,146 [INFO] Resolved core_metadata_collection.yaml\n",
      "2025-10-07 16:44:41,146 [INFO] Resolved all references in schema list.\n",
      "2025-10-07 16:44:41,146 [INFO] Converting schema list to JSON format.\n",
      "2025-10-07 16:44:41,147 [INFO] Converted resolved schema list to JSON format.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key: demographic.yaml\n",
      "key: project.yaml\n",
      "key: acknowledgement.yaml\n",
      "key: medical_history.yaml\n",
      "key: program.yaml\n",
      "key: sample.yaml\n",
      "key: lipidomics_file.yaml\n",
      "key: publication.yaml\n",
      "key: subject.yaml\n",
      "key: core_metadata_collection.yaml\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_list = bundled_schema_to_dict_list('../examples//schema/json/schema_dev.json', return_aux=True)\n",
    "schema_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3013c469",
   "metadata": {},
   "outputs": [],
   "source": [
    "schema_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "38022521",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'project.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'version': None,\n",
       "  'id': 'project',\n",
       "  'title': 'project',\n",
       "  'type': 'object',\n",
       "  'namespace': None,\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Gen3 Compulsary Node',\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['project_id'],\n",
       "  'links': [],\n",
       "  'properties': {'project_id': {'type': 'string',\n",
       "    'description': 'Synthetic_Dataset_1'},\n",
       "   'description': {'type': 'string',\n",
       "    'description': 'Project containing synthetic data'}}},\n",
       " 'sample.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'version': None,\n",
       "  'id': 'sample',\n",
       "  'title': 'sample',\n",
       "  'type': 'object',\n",
       "  'namespace': None,\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Info about sample',\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['sample_id', 'collection_date'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'samples',\n",
       "    'label': None,\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'one_to_many',\n",
       "    'required': True}],\n",
       "  'properties': {'sample_id': {'type': 'string',\n",
       "    'description': 'Sample ID (string)'},\n",
       "   'sample_count': {'type': 'integer',\n",
       "    'description': 'Number of aliquots (integer)'},\n",
       "   'sample_volume': {'type': 'number',\n",
       "    'description': 'Volume in microliters (number/float)'},\n",
       "   'is_viable': {'type': 'boolean',\n",
       "    'description': 'Is the sample viable? (boolean)'},\n",
       "   'collection_date': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'sample_tube_type': {'description': 'Sample tube type (enum)',\n",
       "    'enum': ['EDTA', 'Heparin', 'Citrate']},\n",
       "   'notes': {'type': 'string', 'description': 'Free text notes (string)'},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]}}},\n",
       " 'lipidomics_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'version': None,\n",
       "  'id': 'lipidomics_file',\n",
       "  'title': 'lipidomics_file',\n",
       "  'type': 'object',\n",
       "  'namespace': None,\n",
       "  'category': 'data_file',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Info about lipidomics file',\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['submitter_id', 'type'],\n",
       "  'links': {'exclusive': False,\n",
       "   'required': True,\n",
       "   'subgroup': [{'name': 'samples',\n",
       "     'backref': 'lipidomics_files',\n",
       "     'label': None,\n",
       "     'target_type': 'sample',\n",
       "     'multiplicity': 'one_to_many',\n",
       "     'required': True},\n",
       "    {'name': 'assays',\n",
       "     'backref': 'lipidomics_files',\n",
       "     'label': None,\n",
       "     'target_type': 'assay',\n",
       "     'multiplicity': 'one_to_many',\n",
       "     'required': True},\n",
       "    {'name': 'core_metadata_collections',\n",
       "     'backref': 'lipidomics_files',\n",
       "     'label': None,\n",
       "     'target_type': 'core_metadata_collection',\n",
       "     'multiplicity': 'one_to_one',\n",
       "     'required': True}]},\n",
       "  'properties': {'samples': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'assays': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]}}},\n",
       " 'assay.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'version': None,\n",
       "  'id': 'assay',\n",
       "  'title': 'assay',\n",
       "  'type': 'object',\n",
       "  'namespace': None,\n",
       "  'category': 'data_file',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Info about assay',\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['submitter_id', 'type'],\n",
       "  'links': {'exclusive': False,\n",
       "   'required': True,\n",
       "   'subgroup': [{'name': 'samples',\n",
       "     'backref': 'assays',\n",
       "     'label': None,\n",
       "     'target_type': 'sample',\n",
       "     'multiplicity': 'many_to_one',\n",
       "     'required': True},\n",
       "    {'name': 'core_metadata_collections',\n",
       "     'backref': 'assays',\n",
       "     'label': None,\n",
       "     'target_type': 'core_metadata_collection',\n",
       "     'multiplicity': 'one_to_one',\n",
       "     'required': True}]},\n",
       "  'properties': {'samples': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]}}},\n",
       " 'core_metadata_collection.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'core_metadata_collection',\n",
       "  'title': 'Core Metadata Collection',\n",
       "  'type': 'object',\n",
       "  'namespace': 'https://dcp.bionimbus.org/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Structured description of a collection of several dataset\\n',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'core_metadata_collections',\n",
       "    'label': 'data_from',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['submitter_id', 'type', 'projects'],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': ['string'],\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'contributor': {'description': 'An node responsible for making contributions to the resource. Examples of a Contributor include a person, an organization, or a service. Typically, the name of a Contributor should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'coverage': {'description': 'The spatial or temporal topic of the resource, the spatial applicability of the resource, or the jurisdiction under which the resource is relevant. Spatial topic and spatial applicability may be a named place or a location specified by its geographic coordinates. Temporal topic may be a named period, date, or date range. A jurisdiction may be a named administrative node or a geographic place to which the resource applies. Recommended best practice is to use a controlled vocabulary such as the Thesaurus of Geographic Names [TGN] (http://www.getty.edu/research/tools/vocabulary/tgn/index.html). Where appropriate, named places or time periods can be used in preference to numeric identifiers such as sets of coordinates or date ranges.\\n',\n",
       "    'type': 'string'},\n",
       "   'creator': {'description': 'An node primarily responsible for making the resource. Examples of a Creator include a person, an organization, or a service. Typically, the name of a Creator should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'date': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'description': {'description': 'An account of the resource. Description may include but is not limited to: an abstract, a table of contents, a graphical representation, or a free-text account of the resource.\\n',\n",
       "    'type': 'string'},\n",
       "   'format': {'description': 'The file format, physical medium, or dimensions of the resource. Examples of dimensions include size and duration. Recommended best practice is to use a controlled vocabulary such as the list of Internet Media Types [MIME] (http://www.iana.org/assignments/media-types/).\\n',\n",
       "    'type': 'string'},\n",
       "   'language': {'description': 'A language of the resource. Recommended best practice is to use a controlled vocabulary such as RFC 4646 (http://www.ietf.org/rfc/rfc4646.txt).\\n',\n",
       "    'type': 'string'},\n",
       "   'publisher': {'description': 'An node responsible for making the resource available. Examples of a Publisher include a person, an organization, or a service. Typically, the name of a Publisher should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'relation': {'description': 'A related resource. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\xa0\\n',\n",
       "    'type': 'string'},\n",
       "   'rights': {'description': 'Information about rights held in and over the resource. Typically, rights information includes a statement about various property rights associated with the resource, including intellectual property rights.\\n',\n",
       "    'type': 'string'},\n",
       "   'source': {'description': 'A related resource from which the described resource is derived. The described resource may be derived from the related resource in whole or in part. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\n',\n",
       "    'type': 'string'},\n",
       "   'subject': {'description': 'The topic of the resource. Typically, the subject will be represented using keywords, key phrases, or classification codes. Recommended best practice is to use a controlled vocabulary.\\n',\n",
       "    'type': 'string'},\n",
       "   'title': {'description': 'A name given to the resource. Typically, a Title will be a name by which the resource is formally known.\\n',\n",
       "    'type': 'string'},\n",
       "   'data_type': {'description': 'The nature or genre of the resource. Recommended best practice is to use a controlled vocabulary such as the DCMI Type Vocabulary [DCMITYPE]. To describe the file format, physical medium, or dimensions of the resource, use the Format element.\\n',\n",
       "    'type': 'string'},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]}}}}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from gen3_validator.resolve_schema import ResolveSchema\n",
    "file_input = '../output/test_schema_bundle.json'\n",
    "resolver = ResolveSchema(file_input)\n",
    "resolver.resolve_schema()\n",
    "bundled = resolver.schema_resolved\n",
    "bundled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba777733",
   "metadata": {},
   "outputs": [],
   "source": [
    "schema_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19d3da06",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.converter import format_multiplicity\n",
    "format_multiplicity(multiplicity=\"many_to_one\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "444ebc01",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dbb511d5",
   "metadata": {},
   "source": [
    "# Need to add function in converter that recognises the datetime data input and adds a reference to the _definitions.yaml#/datetime file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38f2ee72",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gen3schemadev.schema.gen3_template import *\n",
    "from gen3schemadev.utils import *\n",
    "from gen3schemadev.schema.input_schema import DataModel\n",
    "from gen3schemadev.converter import *\n",
    "\n",
    "\n",
    "\n",
    "# Loading template and metaschema\n",
    "metaschema = get_metaschema()\n",
    "converter_template = generate_gen3_template(metaschema)\n",
    "\n",
    "# loading input example\n",
    "data = load_yaml('../tests/input_example.yml')\n",
    "validated_model = DataModel.model_validate(data)\n",
    "\n",
    "\n",
    "import json\n",
    "print(json.dumps(construct_props('sample', validated_model), indent=4))\n",
    "\n",
    "\n",
    "# construct_props('sample', validated_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12e03397",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-08 14:19:59,025 [INFO] Initializing DataDictionary with schema path: ./../tests/gen3_schema/schema_dev_fail.json\n",
      "2025-10-08 14:19:59,026 [INFO] Initializing ResolveSchema with schema path: ./../tests/gen3_schema/schema_dev_fail.json\n",
      "2025-10-08 14:19:59,026 [INFO] Starting schema resolution process.\n",
      "2025-10-08 14:19:59,027 [INFO] Reading JSON file from path: ./../tests/gen3_schema/schema_dev_fail.json\n",
      "2025-10-08 14:19:59,028 [INFO] Successfully read JSON schema.\n",
      "2025-10-08 14:19:59,028 [INFO] Retrieving node names from schema.\n",
      "2025-10-08 14:19:59,028 [INFO] Splitting schema into individual node schemas.\n",
      "2025-10-08 14:19:59,028 [INFO] Split schema into individual node schemas.\n",
      "2025-10-08 14:19:59,029 [INFO] Retrieving schema for schema ID: _definitions.yaml\n",
      "2025-10-08 14:19:59,029 [INFO] Retrieved definitions schema.\n",
      "2025-10-08 14:19:59,029 [INFO] Retrieving schema for schema ID: _terms.yaml\n",
      "2025-10-08 14:19:59,029 [INFO] Retrieved terms schema.\n",
      "2025-10-08 14:19:59,030 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,030 [INFO] Resolved references in definitions schema.\n",
      "2025-10-08 14:19:59,030 [INFO] Resolving all references in schema list.\n",
      "2025-10-08 14:19:59,030 [INFO] === Resolving Schema References ===\n",
      "2025-10-08 14:19:59,031 [INFO] Retrieving node names from schema.\n",
      "2025-10-08 14:19:59,031 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,031 [INFO] Resolved demographic.yaml\n",
      "2025-10-08 14:19:59,031 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,031 [INFO] Resolved project.yaml\n",
      "2025-10-08 14:19:59,031 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,032 [INFO] Resolved acknowledgement.yaml\n",
      "2025-10-08 14:19:59,032 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,032 [INFO] Resolved medical_history.yaml\n",
      "2025-10-08 14:19:59,032 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,032 [INFO] Resolved _settings.yaml\n",
      "2025-10-08 14:19:59,033 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,033 [INFO] Resolved program.yaml\n",
      "2025-10-08 14:19:59,033 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,033 [INFO] Resolved sample.yaml\n",
      "2025-10-08 14:19:59,033 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,034 [INFO] Resolved lipidomics_file.yaml\n",
      "2025-10-08 14:19:59,034 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,034 [INFO] Resolved publication.yaml\n",
      "2025-10-08 14:19:59,034 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,034 [INFO] Resolved subject.yaml\n",
      "2025-10-08 14:19:59,034 [INFO] Resolving references in schema.\n",
      "2025-10-08 14:19:59,034 [INFO] Resolved core_metadata_collection.yaml\n",
      "2025-10-08 14:19:59,035 [INFO] Resolved all references in schema list.\n",
      "2025-10-08 14:19:59,035 [INFO] Converting schema list to JSON format.\n",
      "2025-10-08 14:19:59,035 [INFO] Converted resolved schema list to JSON format.\n",
      "2025-10-08 14:19:59,035 [INFO] Validating schema 'demographic' against the Gen3 metaschema.\n",
      "2025-10-08 14:19:59,290 [ERROR] check-jsonschema failed with exit code 1\n",
      "2025-10-08 14:19:59,291 [ERROR] STDOUT: Schema validation errors were encountered.\n",
      "  /Users/harrijh/projects/gen3schemadev/jupyter/.cache/tmphzfwl0ec.json::$.properties.year_birth.type: 'date-time' is not valid under any of the given schemas\n",
      "  Underlying errors caused this.\n",
      "\n",
      "  Best Match:\n",
      "    $.properties.year_birth.type: 'date-time' is not one of ['array', 'boolean', 'integer', 'null', 'number', 'object', 'string']\n",
      "  All Errors:\n",
      "    $.properties.year_birth.type: 'date-time' is not one of ['array', 'boolean', 'integer', 'null', 'number', 'object', 'string']\n",
      "    $.properties.year_birth.type: 'date-time' is not of type 'array'\n",
      "  /Users/harrijh/projects/gen3schemadev/jupyter/.cache/tmphzfwl0ec.json::$.category: 'a_random_category' is not one of ['administrative', 'analysis', 'biospecimen', 'clinical', 'data', 'data_bundle', 'data_file', 'index_file', 'metadata_file', 'notation', 'qc_bundle', 'TBD']\n",
      "  /Users/harrijh/projects/gen3schemadev/jupyter/.cache/tmphzfwl0ec.json::$.links[0]: {'names': 'subjects', 'backref': 'demographics', 'label': 'describes', 'target_type': 'subject', 'multiplicity': 'one_to_one', 'required': True} is not valid under any of the given schemas\n",
      "  Underlying errors caused this.\n",
      "\n",
      "  Best Match:\n",
      "    $.links[0]: Additional properties are not allowed ('names' was unexpected)\n",
      "  All Errors:\n",
      "    $.links[0]: Additional properties are not allowed ('names' was unexpected)\n",
      "    $.links[0]: 'name' is a required property\n",
      "    $.links[0]: Additional properties are not allowed ('backref', 'label', 'multiplicity', 'names', 'target_type' were unexpected)\n",
      "    $.links[0]: 'exclusive' is a required property\n",
      "    $.links[0]: 'subgroup' is a required property\n",
      "  /Users/harrijh/projects/gen3schemadev/jupyter/.cache/tmphzfwl0ec.json::$.properties.year_birth: {'description': 'Year of birth at baseline (YYYY)', 'type': 'date-time'} is not valid under any of the given schemas\n",
      "  Underlying errors caused this.\n",
      "\n",
      "  Best Match:\n",
      "    $.properties.year_birth: {'description': 'Year of birth at baseline (YYYY)', 'type': 'date-time'} is not of type 'string'\n",
      "  Best Deep Match:\n",
      "    $.properties.year_birth.type: 'date-time' is not one of ['string', 'number', 'integer', 'boolean', 'array', 'object']\n",
      "  All Errors:\n",
      "    $.properties.year_birth.type: 'date-time' is not valid under any of the given schemas\n",
      "    $.properties.year_birth.type: 'date-time' is not one of ['string', 'number', 'integer', 'boolean', 'array', 'object']\n",
      "    $.properties.year_birth.type: 'date-time' is not of type 'array'\n",
      "    $.properties.year_birth: {'description': 'Year of birth at baseline (YYYY)', 'type': 'date-time'} is not of type 'string'\n",
      "\n",
      "2025-10-08 14:19:59,291 [ERROR] An unexpected error occurred during metaschema validation.\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/harrijh/projects/gen3schemadev/src/gen3schemadev/validators/metaschema_validator.py\", line 103, in validate_schema_with_metaschema\n",
      "    raise RuntimeError(\n",
      "RuntimeError: check-jsonschema validation failed for schema 'demographic'. See logs for details.\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "check-jsonschema validation failed for schema 'demographic'. See logs for details.",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mRuntimeError\u001b[39m                              Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 12\u001b[39m\n\u001b[32m      9\u001b[39m schema_list = bundled_schema_to_list_dict(resolve_schema)\n\u001b[32m     11\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m schema \u001b[38;5;129;01min\u001b[39;00m schema_list:\n\u001b[32m---> \u001b[39m\u001b[32m12\u001b[39m     \u001b[43mvalidate_schema_with_metaschema\u001b[49m\u001b[43m(\u001b[49m\u001b[43mschema\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetaschema\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmetaschema\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/projects/gen3schemadev/src/gen3schemadev/validators/metaschema_validator.py:103\u001b[39m, in \u001b[36mvalidate_schema_with_metaschema\u001b[39m\u001b[34m(schema, metaschema, verbose)\u001b[39m\n\u001b[32m    101\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m completed_process.stderr:\n\u001b[32m    102\u001b[39m         logger.error(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mSTDERR: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcompleted_process.stderr\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m103\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[32m    104\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mcheck-jsonschema validation failed for schema \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mschema.get(\u001b[33m'\u001b[39m\u001b[33mid\u001b[39m\u001b[33m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[33m'\u001b[39m\u001b[33m<no id>\u001b[39m\u001b[33m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m. \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    105\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mSee logs for details.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    106\u001b[39m     )\n\u001b[32m    107\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    108\u001b[39m     logger.info(\n\u001b[32m    109\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mSchema \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mschema.get(\u001b[33m'\u001b[39m\u001b[33mid\u001b[39m\u001b[33m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[33m'\u001b[39m\u001b[33m<no id>\u001b[39m\u001b[33m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m successfully validated against metaschema.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    110\u001b[39m     )\n",
      "\u001b[31mRuntimeError\u001b[39m: check-jsonschema validation failed for schema 'demographic'. See logs for details."
     ]
    }
   ],
   "source": [
    "from gen3schemadev.schema.gen3_template import *\n",
    "from gen3schemadev.utils import bundled_schema_to_list_dict\n",
    "from gen3schemadev.validators.metaschema_validator import validate_schema_with_metaschema\n",
    "import os\n",
    "\n",
    "metaschema = get_metaschema()\n",
    "schema_file = os.path.join('.', '../tests/gen3_schema/schema_dev_fail.json')\n",
    "resolve_schema = resolve_schema(schema_path=schema_file)\n",
    "schema_list = bundled_schema_to_list_dict(resolve_schema)\n",
    "\n",
    "for schema in schema_list:\n",
    "    validate_schema_with_metaschema(schema, metaschema=metaschema, verbose=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7112c66d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       " 'id': 'demographic',\n",
       " 'title': 'Demographic',\n",
       " 'type': 'object',\n",
       " 'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       " 'category': 'a_random_category',\n",
       " 'program': '*',\n",
       " 'project': '*',\n",
       " 'description': 'Data for the characterization of the patient by means of segementing the population (e.g. characterization by age, sex, or race).',\n",
       " 'additionalProperties': False,\n",
       " 'submittable': True,\n",
       " 'validators': None,\n",
       " 'systemProperties': ['id',\n",
       "  'project_id',\n",
       "  'state',\n",
       "  'created_datetime',\n",
       "  'updated_datetime'],\n",
       " 'links': [{'names': 'subjects',\n",
       "   'backref': 'demographics',\n",
       "   'label': 'describes',\n",
       "   'target_type': 'subject',\n",
       "   'multiplicity': 'one_to_one',\n",
       "   'required': True}],\n",
       " 'required': ['type', 'submitter_id', 'subjects'],\n",
       " 'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       " 'properties': {'type': {'type': 'string'},\n",
       "  'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "    'termDef': {'term': 'Universally Unique Identifier',\n",
       "     'source': 'NCIt',\n",
       "     'cde_id': 'C54100',\n",
       "     'cde_version': None,\n",
       "     'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "   'type': 'string',\n",
       "   'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "   'systemAlias': 'node_id'},\n",
       "  'submitter_id': {'type': 'string',\n",
       "   'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "  'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "   'default': 'validated',\n",
       "   'downloadable': ['uploaded',\n",
       "    'md5summed',\n",
       "    'validating',\n",
       "    'validated',\n",
       "    'error',\n",
       "    'invalid',\n",
       "    'released'],\n",
       "   'public': ['live'],\n",
       "   'oneOf': [{'enum': ['uploading',\n",
       "      'uploaded',\n",
       "      'md5summing',\n",
       "      'md5summed',\n",
       "      'validating',\n",
       "      'error',\n",
       "      'invalid',\n",
       "      'suppressed',\n",
       "      'redacted',\n",
       "      'live']},\n",
       "    {'enum': ['validated', 'submitted', 'released']}]},\n",
       "  'project_id': {'type': 'string',\n",
       "   'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "  'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "    {'type': 'null'}],\n",
       "   'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "  'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "    {'type': 'null'}],\n",
       "   'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "  'subjects': {'anyOf': [{'type': 'array',\n",
       "     'items': {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}},\n",
       "      'minItems': 1}},\n",
       "    {'type': 'object',\n",
       "     'additionalProperties': True,\n",
       "     'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "        'termDef': {'term': 'Universally Unique Identifier',\n",
       "         'source': 'NCIt',\n",
       "         'cde_id': 'C54100',\n",
       "         'cde_version': None,\n",
       "         'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "       'type': 'string',\n",
       "       'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "      'submitter_id': {'type': 'string'}}}]},\n",
       "  'year_birth': {'description': 'Year of birth at baseline (YYYY)',\n",
       "   'type': 'date-time'},\n",
       "  'month_birth': {'description': 'Month of birth at baseline (MM)',\n",
       "   'type': 'integer',\n",
       "   'location': 'australia'}},\n",
       " 'an_extra_property': 'extra_value'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demographic_schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5426dc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-07 17:36:37,877 [INFO] Initializing DataDictionary with schema path: ./../tests/gen3_schema/schema_dev_pass.json\n",
      "2025-10-07 17:36:37,878 [INFO] Initializing ResolveSchema with schema path: ./../tests/gen3_schema/schema_dev_pass.json\n",
      "2025-10-07 17:36:37,878 [INFO] Starting schema resolution process.\n",
      "2025-10-07 17:36:37,878 [INFO] Reading JSON file from path: ./../tests/gen3_schema/schema_dev_pass.json\n",
      "2025-10-07 17:36:37,879 [INFO] Successfully read JSON schema.\n",
      "2025-10-07 17:36:37,879 [INFO] Retrieving node names from schema.\n",
      "2025-10-07 17:36:37,879 [INFO] Splitting schema into individual node schemas.\n",
      "2025-10-07 17:36:37,880 [INFO] Split schema into individual node schemas.\n",
      "2025-10-07 17:36:37,880 [INFO] Retrieving schema for schema ID: _definitions.yaml\n",
      "2025-10-07 17:36:37,880 [INFO] Retrieved definitions schema.\n",
      "2025-10-07 17:36:37,880 [INFO] Retrieving schema for schema ID: _terms.yaml\n",
      "2025-10-07 17:36:37,880 [INFO] Retrieved terms schema.\n",
      "2025-10-07 17:36:37,880 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,880 [INFO] Resolved references in definitions schema.\n",
      "2025-10-07 17:36:37,881 [INFO] Resolving all references in schema list.\n",
      "2025-10-07 17:36:37,881 [INFO] === Resolving Schema References ===\n",
      "2025-10-07 17:36:37,881 [INFO] Retrieving node names from schema.\n",
      "2025-10-07 17:36:37,881 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,881 [INFO] Resolved demographic.yaml\n",
      "2025-10-07 17:36:37,881 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,882 [INFO] Resolved project.yaml\n",
      "2025-10-07 17:36:37,882 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,882 [INFO] Resolved acknowledgement.yaml\n",
      "2025-10-07 17:36:37,882 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,882 [INFO] Resolved medical_history.yaml\n",
      "2025-10-07 17:36:37,882 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,882 [INFO] Resolved _settings.yaml\n",
      "2025-10-07 17:36:37,883 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,883 [INFO] Resolved program.yaml\n",
      "2025-10-07 17:36:37,883 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,883 [INFO] Resolved sample.yaml\n",
      "2025-10-07 17:36:37,883 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,883 [INFO] Resolved lipidomics_file.yaml\n",
      "2025-10-07 17:36:37,884 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,884 [INFO] Resolved publication.yaml\n",
      "2025-10-07 17:36:37,884 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,884 [INFO] Resolved subject.yaml\n",
      "2025-10-07 17:36:37,884 [INFO] Resolving references in schema.\n",
      "2025-10-07 17:36:37,884 [INFO] Resolved core_metadata_collection.yaml\n",
      "2025-10-07 17:36:37,885 [INFO] Resolved all references in schema list.\n",
      "2025-10-07 17:36:37,885 [INFO] Converting schema list to JSON format.\n",
      "2025-10-07 17:36:37,885 [INFO] Converted resolved schema list to JSON format.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from gen3schemadev.utils import *\n",
    "schema_file = os.path.join('.', '../tests/gen3_schema/schema_dev_pass.json')\n",
    "resolved_schema = resolve_schema(schema_file)\n",
    "len(resolved_schema)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0729d64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'demographic',\n",
       "  'title': 'Demographic',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Data for the characterization of the patient by means of segementing the population (e.g. characterization by age, sex, or race).',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'demographics',\n",
       "    'label': 'describes',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'subjects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'year_birth': {'description': 'Year of birth at baseline (YYYY)',\n",
       "    'type': 'integer'},\n",
       "   'month_birth': {'description': 'Month of birth at baseline (MM)',\n",
       "    'type': 'integer'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'project',\n",
       "  'title': 'Project',\n",
       "  'type': 'object',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'category': 'administrative',\n",
       "  'description': 'The study the data is coming from',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'state',\n",
       "   'released',\n",
       "   'releasable',\n",
       "   'intended_release_date'],\n",
       "  'required': ['code', 'name', 'programs', 'dbgap_accession_number'],\n",
       "  'uniqueKeys': [['id'], ['code']],\n",
       "  'links': [{'name': 'programs',\n",
       "    'backref': 'projects',\n",
       "    'label': 'member_of',\n",
       "    'target_type': 'program',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'constraints': None,\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id',\n",
       "    'description': 'UUID for the project.'},\n",
       "   'name': {'type': 'string',\n",
       "    'description': 'Display name/brief description for the project.'},\n",
       "   'code': {'type': 'string',\n",
       "    'description': 'Unique identifier for the project.'},\n",
       "   'date_collected': {'description': 'The date or date range in which the project data was collected.',\n",
       "    'type': 'string'},\n",
       "   'availability_type': {'description': 'Is the project open or restricted?',\n",
       "    'enum': ['Open', 'Restricted']},\n",
       "   'availability_mechanism': {'description': 'Mechanism by which the project will be made avilable.',\n",
       "    'type': 'string'},\n",
       "   'support_source': {'description': 'The name of source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'support_id': {'description': 'The ID of the source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'programs': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}],\n",
       "    'description': 'Indicates that the Study is logically part of the indicated Study.\\n'},\n",
       "   'state': {'description': 'The possible states a project can be in.  All but `open` are\\nequivalent to some type of locked state.\\n',\n",
       "    'default': 'open',\n",
       "    'enum': ['open', 'review', 'submitted', 'processing', 'closed', 'legacy']},\n",
       "   'released': {'description': 'To release a project is to tell the GDC to include all submitted\\nnodes in the next GDC index.\\n',\n",
       "    'default': False,\n",
       "    'type': 'boolean'},\n",
       "   'releasable': {'description': 'A project can only be released by the user when `releasable` is true.\\n',\n",
       "    'default': False,\n",
       "    'type': 'boolean'},\n",
       "   'intended_release_date': {'description': \"Tracks a Project's intended release date.\",\n",
       "    'type': 'string',\n",
       "    'format': 'date-time'},\n",
       "   'dbgap_accession_number': {'type': 'string',\n",
       "    'description': 'The dbgap accession number provided for the project. (CMG)'},\n",
       "   'authz': {'description': 'The authz string (example, [\"/programs/Program.name/projects/Project.code\"] ).',\n",
       "    'type': 'array',\n",
       "    'items': {'type': 'string'}},\n",
       "   'short_name': {'description': 'The shorthand name for the project.',\n",
       "    'type': 'string'},\n",
       "   'full_name': {'description': 'The full name for the project.',\n",
       "    'type': 'string'},\n",
       "   'project_description': {'description': 'The description for the project.',\n",
       "    'type': 'string'},\n",
       "   'project_registration': {'description': 'External source from which the identifier included in study_id originates',\n",
       "    'type': 'string'},\n",
       "   'consent_codes': {'description': 'Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. \\\\n\\n        Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html',\n",
       "    'type': 'array'},\n",
       "   'data_access_url': {'description': 'A URL link that takes you to a page that explains how to gain access to the project',\n",
       "    'type': 'string'}},\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/'},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'acknowledgement',\n",
       "  'title': 'Acknowledgement',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Acknowledgement of an individual or group involved in a project.',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'acknowledgements',\n",
       "    'label': 'contribute_to',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'medical_history',\n",
       "  'title': 'Medical History',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Medical history of the participant',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'medical_histories',\n",
       "    'label': 'describes',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'subjects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'diabetes_type': {'description': 'diabetes diagnosed by fasting blood glucose >=7 mmol/l or tx AHA or 2 hour blood glucose >=11.1 mmol/l.',\n",
       "    'termDef': [{'term': 'Diabetes mellitus',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0000819',\n",
       "      'term_version': '2021-10-10'}],\n",
       "    'enum': ['IGT', 'KDM', 'IFG', 'NDM', 'NGT'],\n",
       "    'enumDef': [{'enumeration': 'IGT',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0040270'},\n",
       "     {'enumeration': 'NDM', 'source': 'SNOMED', 'term_id': '870528001'},\n",
       "     {'enumeration': 'NGT', 'source': 'SNOMED', 'term_id': '166926006'}]},\n",
       "   'hypertension': {'description': 'Whether the participant has Hypertension',\n",
       "    'enum': ['yes, measured or on treatment',\n",
       "     'yes, self-reported',\n",
       "     'no',\n",
       "     'not reported',\n",
       "     'not collected']},\n",
       "   'hypertension_measurement_type': {'description': 'Whether the hypertension was measured at recruitment/patients on hypertension treatment or self-reported medical diagnosis',\n",
       "    'enum': ['self-reported', 'measured']},\n",
       "   'diabetes': {'description': 'Self-reported diabetes in the participant',\n",
       "    'termDef': [{'term': 'Diabetes mellitus',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0000819'}],\n",
       "    'enum': ['yes', 'no', 'unsure', 'not reported', 'not collected']}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'program',\n",
       "  'title': 'Program',\n",
       "  'type': 'object',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'A broad framework of goals to be achieved. (NCIt C52647)\\n',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': False,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id'],\n",
       "  'required': ['name', 'dbgap_accession_number'],\n",
       "  'uniqueKeys': [['id'], ['name']],\n",
       "  'links': [],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'name': {'type': 'string',\n",
       "    'description': 'Full name/title of the program.'},\n",
       "   'dbgap_accession_number': {'type': 'string',\n",
       "    'description': 'The dbgap accession number provided for the program.'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'sample',\n",
       "  'title': 'Sample',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'biospecimen',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': \"Biospecimen information that links subjects to samples including sample's provider and source.\",\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'samples',\n",
       "    'label': 'taken_from',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'subjects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'sample_provider': {'description': 'The name of collaborating institute that provided the sample',\n",
       "    'enum': ['Baker', 'USYD', 'UMELB', 'UQ']},\n",
       "   'sample_source': {'description': 'Uberon identifier, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG, CCDG)',\n",
       "    'type': 'string',\n",
       "    'pattern': '^UBERON:[0-9]{7}$'},\n",
       "   'sample_type': {'description': 'A string saying the type of sample',\n",
       "    'type': 'string'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'lipidomics_file',\n",
       "  'title': 'Lipidomics File',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'data_file',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Data file containing lipidomics data',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'exclusive': False,\n",
       "    'required': True,\n",
       "    'subgroup': [{'name': 'samples',\n",
       "      'backref': 'lipidomics_files',\n",
       "      'label': 'data_from',\n",
       "      'target_type': 'sample',\n",
       "      'multiplicity': 'many_to_many',\n",
       "      'required': False},\n",
       "     {'name': 'core_metadata_collection',\n",
       "      'backref': 'lipidomics_files',\n",
       "      'label': 'describes',\n",
       "      'target_type': 'core_metadata_collection',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'required': True}]}],\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'core_metadata_collection',\n",
       "   'data_type',\n",
       "   'data_format',\n",
       "   'data_category'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'file_name': {'type': 'string',\n",
       "    'term': {'description': 'The name (or part of a name) of a file (of any type).\\n'}},\n",
       "   'file_size': {'type': 'integer',\n",
       "    'term': {'description': 'The size of the data file (object) in bytes.\\n'}},\n",
       "   'file_format': {'type': 'string',\n",
       "    'term': {'description': 'The format of the data file object.\\n'}},\n",
       "   'md5sum': {'type': 'string',\n",
       "    'term': {'description': \"The 128-bit hash value expressed as a 32 digit hexadecimal number used as a file's digital fingerprint.\\n\"},\n",
       "    'pattern': '^[a-f0-9]{32}$'},\n",
       "   'object_id': {'type': 'string',\n",
       "    'description': 'The GUID of the object in the index service.'},\n",
       "   'file_state': {'term': {'description': 'The current state of the data file object.\\n'},\n",
       "    'default': 'registered',\n",
       "    'enum': ['registered',\n",
       "     'uploading',\n",
       "     'uploaded',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'submitted',\n",
       "     'processing',\n",
       "     'processed',\n",
       "     'released',\n",
       "     'error']},\n",
       "   'error_type': {'term': {'description': 'Type of error for the data file object.\\n'},\n",
       "    'enum': ['file_size', 'file_format', 'md5sum']},\n",
       "   'ga4gh_drs_uri': {'type': 'string',\n",
       "    'term': {'description': 'DRS URI as defined by GA4GH DRS spec for pointers to file objects.\\n'}},\n",
       "   'samples': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'core_metadata_collection': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'data_type': {'description': 'The type of data in this data file',\n",
       "    'enum': ['TIC/XIC', 'MS', 'MS/MS', 'Image']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['csv', 'tsv', 'xls', 'xlsx']},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'cv': {'description': 'Coefficient of variation (%)', 'type': 'number'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'publication',\n",
       "  'title': 'Publication',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Publication for a project',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'publications',\n",
       "    'label': 'refers_to',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'subject',\n",
       "  'title': 'Subject',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'An individual participant in the study with baseline measurements.',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'subjects',\n",
       "    'label': 'part_of',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects', 'baseline_year'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]},\n",
       "   'consent_codes': {'description': 'Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. \\\\n\\n        Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html',\n",
       "    'type': 'array'},\n",
       "   'baseline_year': {'description': 'year when baseline measurements were taken',\n",
       "    'type': 'string'},\n",
       "   'cohort_id': {'description': 'Identifier for the cohort the subject is a part of',\n",
       "    'type': 'string'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'core_metadata_collection',\n",
       "  'title': 'Core Metadata Collection',\n",
       "  'type': 'object',\n",
       "  'namespace': 'https://dcp.bionimbus.org/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Structured description of a collection of several dataset\\n',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'core_metadata_collections',\n",
       "    'label': 'data_from',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['submitter_id', 'type', 'projects'],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'contributor': {'description': 'An node responsible for making contributions to the resource. Examples of a Contributor include a person, an organization, or a service. Typically, the name of a Contributor should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'coverage': {'description': 'The spatial or temporal topic of the resource, the spatial applicability of the resource, or the jurisdiction under which the resource is relevant. Spatial topic and spatial applicability may be a named place or a location specified by its geographic coordinates. Temporal topic may be a named period, date, or date range. A jurisdiction may be a named administrative node or a geographic place to which the resource applies. Recommended best practice is to use a controlled vocabulary such as the Thesaurus of Geographic Names [TGN] (http://www.getty.edu/research/tools/vocabulary/tgn/index.html). Where appropriate, named places or time periods can be used in preference to numeric identifiers such as sets of coordinates or date ranges.\\n',\n",
       "    'type': 'string'},\n",
       "   'creator': {'description': 'An node primarily responsible for making the resource. Examples of a Creator include a person, an organization, or a service. Typically, the name of a Creator should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'date': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'description': {'description': 'An account of the resource. Description may include but is not limited to: an abstract, a table of contents, a graphical representation, or a free-text account of the resource.\\n',\n",
       "    'type': 'string'},\n",
       "   'format': {'description': 'The file format, physical medium, or dimensions of the resource. Examples of dimensions include size and duration. Recommended best practice is to use a controlled vocabulary such as the list of Internet Media Types [MIME] (http://www.iana.org/assignments/media-types/).\\n',\n",
       "    'type': 'string'},\n",
       "   'language': {'description': 'A language of the resource. Recommended best practice is to use a controlled vocabulary such as RFC 4646 (http://www.ietf.org/rfc/rfc4646.txt).\\n',\n",
       "    'type': 'string'},\n",
       "   'publisher': {'description': 'An node responsible for making the resource available. Examples of a Publisher include a person, an organization, or a service. Typically, the name of a Publisher should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'relation': {'description': 'A related resource. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\xa0\\n',\n",
       "    'type': 'string'},\n",
       "   'rights': {'description': 'Information about rights held in and over the resource. Typically, rights information includes a statement about various property rights associated with the resource, including intellectual property rights.\\n',\n",
       "    'type': 'string'},\n",
       "   'source': {'description': 'A related resource from which the described resource is derived. The described resource may be derived from the related resource in whole or in part. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\n',\n",
       "    'type': 'string'},\n",
       "   'subject': {'description': 'The topic of the resource. Typically, the subject will be represented using keywords, key phrases, or classification codes. Recommended best practice is to use a controlled vocabulary.\\n',\n",
       "    'type': 'string'},\n",
       "   'title': {'description': 'A name given to the resource. Typically, a Title will be a name by which the resource is formally known.\\n',\n",
       "    'type': 'string'},\n",
       "   'data_type': {'description': 'The nature or genre of the resource. Recommended best practice is to use a controlled vocabulary such as the DCMI Type Vocabulary [DCMITYPE]. To describe the file format, physical medium, or dimensions of the resource, use the Format element.\\n',\n",
       "    'type': 'string'},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]}}}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resolved_schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8bb4f632",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'demographic',\n",
       "  'title': 'Demographic',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Data for the characterization of the patient by means of segementing the population (e.g. characterization by age, sex, or race).',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'demographics',\n",
       "    'label': 'describes',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'subjects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'year_birth': {'description': 'Year of birth at baseline (YYYY)',\n",
       "    'type': 'integer'},\n",
       "   'month_birth': {'description': 'Month of birth at baseline (MM)',\n",
       "    'type': 'integer'}}}]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demographic = [schema for schema in schema_list if schema['id'] == 'demographic']\n",
    "demographic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "82884264",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'project',\n",
       "  'title': 'Project',\n",
       "  'type': 'object',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'category': 'administrative',\n",
       "  'description': 'The study the data is coming from',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'state',\n",
       "   'released',\n",
       "   'releasable',\n",
       "   'intended_release_date'],\n",
       "  'required': ['code', 'name', 'programs', 'dbgap_accession_number'],\n",
       "  'uniqueKeys': [['id'], ['code']],\n",
       "  'links': [{'name': 'programs',\n",
       "    'backref': 'projects',\n",
       "    'label': 'member_of',\n",
       "    'target_type': 'program',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'constraints': None,\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id',\n",
       "    'description': 'UUID for the project.'},\n",
       "   'name': {'type': 'string',\n",
       "    'description': 'Display name/brief description for the project.'},\n",
       "   'code': {'type': 'string',\n",
       "    'description': 'Unique identifier for the project.'},\n",
       "   'date_collected': {'description': 'The date or date range in which the project data was collected.',\n",
       "    'type': 'string'},\n",
       "   'availability_type': {'description': 'Is the project open or restricted?',\n",
       "    'enum': ['Open', 'Restricted']},\n",
       "   'availability_mechanism': {'description': 'Mechanism by which the project will be made avilable.',\n",
       "    'type': 'string'},\n",
       "   'support_source': {'description': 'The name of source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'support_id': {'description': 'The ID of the source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'programs': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}],\n",
       "    'description': 'Indicates that the Study is logically part of the indicated Study.\\n'},\n",
       "   'state': {'description': 'The possible states a project can be in.  All but `open` are\\nequivalent to some type of locked state.\\n',\n",
       "    'default': 'open',\n",
       "    'enum': ['open', 'review', 'submitted', 'processing', 'closed', 'legacy']},\n",
       "   'released': {'description': 'To release a project is to tell the GDC to include all submitted\\nnodes in the next GDC index.\\n',\n",
       "    'default': False,\n",
       "    'type': 'boolean'},\n",
       "   'releasable': {'description': 'A project can only be released by the user when `releasable` is true.\\n',\n",
       "    'default': False,\n",
       "    'type': 'boolean'},\n",
       "   'intended_release_date': {'description': \"Tracks a Project's intended release date.\",\n",
       "    'type': 'string',\n",
       "    'format': 'date-time'},\n",
       "   'dbgap_accession_number': {'type': 'string',\n",
       "    'description': 'The dbgap accession number provided for the project. (CMG)'},\n",
       "   'authz': {'description': 'The authz string (example, [\"/programs/Program.name/projects/Project.code\"] ).',\n",
       "    'type': 'array',\n",
       "    'items': {'type': 'string'}},\n",
       "   'short_name': {'description': 'The shorthand name for the project.',\n",
       "    'type': 'string'},\n",
       "   'full_name': {'description': 'The full name for the project.',\n",
       "    'type': 'string'},\n",
       "   'project_description': {'description': 'The description for the project.',\n",
       "    'type': 'string'},\n",
       "   'project_registration': {'description': 'External source from which the identifier included in study_id originates',\n",
       "    'type': 'string'},\n",
       "   'consent_codes': {'description': 'Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. \\\\n\\n        Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html',\n",
       "    'type': 'array'},\n",
       "   'data_access_url': {'description': 'A URL link that takes you to a page that explains how to gain access to the project',\n",
       "    'type': 'string'}},\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/'},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'acknowledgement',\n",
       "  'title': 'Acknowledgement',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Acknowledgement of an individual or group involved in a project.',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'acknowledgements',\n",
       "    'label': 'contribute_to',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'medical_history',\n",
       "  'title': 'Medical History',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Medical history of the participant',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'medical_histories',\n",
       "    'label': 'describes',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'subjects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'diabetes_type': {'description': 'diabetes diagnosed by fasting blood glucose >=7 mmol/l or tx AHA or 2 hour blood glucose >=11.1 mmol/l.',\n",
       "    'termDef': [{'term': 'Diabetes mellitus',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0000819',\n",
       "      'term_version': '2021-10-10'}],\n",
       "    'enum': ['IGT', 'KDM', 'IFG', 'NDM', 'NGT'],\n",
       "    'enumDef': [{'enumeration': 'IGT',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0040270'},\n",
       "     {'enumeration': 'NDM', 'source': 'SNOMED', 'term_id': '870528001'},\n",
       "     {'enumeration': 'NGT', 'source': 'SNOMED', 'term_id': '166926006'}]},\n",
       "   'hypertension': {'description': 'Whether the participant has Hypertension',\n",
       "    'enum': ['yes, measured or on treatment',\n",
       "     'yes, self-reported',\n",
       "     'no',\n",
       "     'not reported',\n",
       "     'not collected']},\n",
       "   'hypertension_measurement_type': {'description': 'Whether the hypertension was measured at recruitment/patients on hypertension treatment or self-reported medical diagnosis',\n",
       "    'enum': ['self-reported', 'measured']},\n",
       "   'diabetes': {'description': 'Self-reported diabetes in the participant',\n",
       "    'termDef': [{'term': 'Diabetes mellitus',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0000819'}],\n",
       "    'enum': ['yes', 'no', 'unsure', 'not reported', 'not collected']}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'program',\n",
       "  'title': 'Program',\n",
       "  'type': 'object',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'A broad framework of goals to be achieved. (NCIt C52647)\\n',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': False,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id'],\n",
       "  'required': ['name', 'dbgap_accession_number'],\n",
       "  'uniqueKeys': [['id'], ['name']],\n",
       "  'links': [],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'name': {'type': 'string',\n",
       "    'description': 'Full name/title of the program.'},\n",
       "   'dbgap_accession_number': {'type': 'string',\n",
       "    'description': 'The dbgap accession number provided for the program.'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'sample',\n",
       "  'title': 'Sample',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'biospecimen',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': \"Biospecimen information that links subjects to samples including sample's provider and source.\",\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'samples',\n",
       "    'label': 'taken_from',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'subjects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'sample_provider': {'description': 'The name of collaborating institute that provided the sample',\n",
       "    'enum': ['Baker', 'USYD', 'UMELB', 'UQ']},\n",
       "   'sample_source': {'description': 'Uberon identifier, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG, CCDG)',\n",
       "    'type': 'string',\n",
       "    'pattern': '^UBERON:[0-9]{7}$'},\n",
       "   'sample_type': {'description': 'A string saying the type of sample',\n",
       "    'type': 'string'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'lipidomics_file',\n",
       "  'title': 'Lipidomics File',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'data_file',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Data file containing lipidomics data',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'exclusive': False,\n",
       "    'required': True,\n",
       "    'subgroup': [{'name': 'samples',\n",
       "      'backref': 'lipidomics_files',\n",
       "      'label': 'data_from',\n",
       "      'target_type': 'sample',\n",
       "      'multiplicity': 'many_to_many',\n",
       "      'required': False},\n",
       "     {'name': 'core_metadata_collection',\n",
       "      'backref': 'lipidomics_files',\n",
       "      'label': 'describes',\n",
       "      'target_type': 'core_metadata_collection',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'required': True}]}],\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'core_metadata_collection',\n",
       "   'data_type',\n",
       "   'data_format',\n",
       "   'data_category'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'file_name': {'type': 'string',\n",
       "    'term': {'description': 'The name (or part of a name) of a file (of any type).\\n'}},\n",
       "   'file_size': {'type': 'integer',\n",
       "    'term': {'description': 'The size of the data file (object) in bytes.\\n'}},\n",
       "   'file_format': {'type': 'string',\n",
       "    'term': {'description': 'The format of the data file object.\\n'}},\n",
       "   'md5sum': {'type': 'string',\n",
       "    'term': {'description': \"The 128-bit hash value expressed as a 32 digit hexadecimal number used as a file's digital fingerprint.\\n\"},\n",
       "    'pattern': '^[a-f0-9]{32}$'},\n",
       "   'object_id': {'type': 'string',\n",
       "    'description': 'The GUID of the object in the index service.'},\n",
       "   'file_state': {'term': {'description': 'The current state of the data file object.\\n'},\n",
       "    'default': 'registered',\n",
       "    'enum': ['registered',\n",
       "     'uploading',\n",
       "     'uploaded',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'submitted',\n",
       "     'processing',\n",
       "     'processed',\n",
       "     'released',\n",
       "     'error']},\n",
       "   'error_type': {'term': {'description': 'Type of error for the data file object.\\n'},\n",
       "    'enum': ['file_size', 'file_format', 'md5sum']},\n",
       "   'ga4gh_drs_uri': {'type': 'string',\n",
       "    'term': {'description': 'DRS URI as defined by GA4GH DRS spec for pointers to file objects.\\n'}},\n",
       "   'samples': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'core_metadata_collection': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'submitter_id': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'submitter_id': {'type': 'string'}}}]},\n",
       "   'data_type': {'description': 'The type of data in this data file',\n",
       "    'enum': ['TIC/XIC', 'MS', 'MS/MS', 'Image']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['csv', 'tsv', 'xls', 'xlsx']},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'cv': {'description': 'Coefficient of variation (%)', 'type': 'number'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'publication',\n",
       "  'title': 'Publication',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Publication for a project',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'publications',\n",
       "    'label': 'refers_to',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'subject',\n",
       "  'title': 'Subject',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'An individual participant in the study with baseline measurements.',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'subjects',\n",
       "    'label': 'part_of',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects', 'baseline_year'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]},\n",
       "   'consent_codes': {'description': 'Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. \\\\n\\n        Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html',\n",
       "    'type': 'array'},\n",
       "   'baseline_year': {'description': 'year when baseline measurements were taken',\n",
       "    'type': 'string'},\n",
       "   'cohort_id': {'description': 'Identifier for the cohort the subject is a part of',\n",
       "    'type': 'string'}}},\n",
       " {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'core_metadata_collection',\n",
       "  'title': 'Core Metadata Collection',\n",
       "  'type': 'object',\n",
       "  'namespace': 'https://dcp.bionimbus.org/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Structured description of a collection of several dataset\\n',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'core_metadata_collections',\n",
       "    'label': 'data_from',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['submitter_id', 'type', 'projects'],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "     'termDef': {'term': 'Universally Unique Identifier',\n",
       "      'source': 'NCIt',\n",
       "      'cde_id': 'C54100',\n",
       "      'cde_version': None,\n",
       "      'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "    'type': 'string',\n",
       "    'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$',\n",
       "    'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'term': {'description': 'The current state of the object.\\n'},\n",
       "    'default': 'validated',\n",
       "    'downloadable': ['uploaded',\n",
       "     'md5summed',\n",
       "     'validating',\n",
       "     'validated',\n",
       "     'error',\n",
       "     'invalid',\n",
       "     'released'],\n",
       "    'public': ['live'],\n",
       "    'oneOf': [{'enum': ['uploading',\n",
       "       'uploaded',\n",
       "       'md5summing',\n",
       "       'md5summed',\n",
       "       'validating',\n",
       "       'error',\n",
       "       'invalid',\n",
       "       'suppressed',\n",
       "       'redacted',\n",
       "       'live']},\n",
       "     {'enum': ['validated', 'submitted', 'released']}]},\n",
       "   'project_id': {'type': 'string',\n",
       "    'term': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'}},\n",
       "   'created_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'updated_datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'contributor': {'description': 'An node responsible for making contributions to the resource. Examples of a Contributor include a person, an organization, or a service. Typically, the name of a Contributor should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'coverage': {'description': 'The spatial or temporal topic of the resource, the spatial applicability of the resource, or the jurisdiction under which the resource is relevant. Spatial topic and spatial applicability may be a named place or a location specified by its geographic coordinates. Temporal topic may be a named period, date, or date range. A jurisdiction may be a named administrative node or a geographic place to which the resource applies. Recommended best practice is to use a controlled vocabulary such as the Thesaurus of Geographic Names [TGN] (http://www.getty.edu/research/tools/vocabulary/tgn/index.html). Where appropriate, named places or time periods can be used in preference to numeric identifiers such as sets of coordinates or date ranges.\\n',\n",
       "    'type': 'string'},\n",
       "   'creator': {'description': 'An node primarily responsible for making the resource. Examples of a Creator include a person, an organization, or a service. Typically, the name of a Creator should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'date': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "     {'type': 'null'}],\n",
       "    'term': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'}},\n",
       "   'description': {'description': 'An account of the resource. Description may include but is not limited to: an abstract, a table of contents, a graphical representation, or a free-text account of the resource.\\n',\n",
       "    'type': 'string'},\n",
       "   'format': {'description': 'The file format, physical medium, or dimensions of the resource. Examples of dimensions include size and duration. Recommended best practice is to use a controlled vocabulary such as the list of Internet Media Types [MIME] (http://www.iana.org/assignments/media-types/).\\n',\n",
       "    'type': 'string'},\n",
       "   'language': {'description': 'A language of the resource. Recommended best practice is to use a controlled vocabulary such as RFC 4646 (http://www.ietf.org/rfc/rfc4646.txt).\\n',\n",
       "    'type': 'string'},\n",
       "   'publisher': {'description': 'An node responsible for making the resource available. Examples of a Publisher include a person, an organization, or a service. Typically, the name of a Publisher should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'relation': {'description': 'A related resource. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\xa0\\n',\n",
       "    'type': 'string'},\n",
       "   'rights': {'description': 'Information about rights held in and over the resource. Typically, rights information includes a statement about various property rights associated with the resource, including intellectual property rights.\\n',\n",
       "    'type': 'string'},\n",
       "   'source': {'description': 'A related resource from which the described resource is derived. The described resource may be derived from the related resource in whole or in part. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\n',\n",
       "    'type': 'string'},\n",
       "   'subject': {'description': 'The topic of the resource. Typically, the subject will be represented using keywords, key phrases, or classification codes. Recommended best practice is to use a controlled vocabulary.\\n',\n",
       "    'type': 'string'},\n",
       "   'title': {'description': 'A name given to the resource. Typically, a Title will be a name by which the resource is formally known.\\n',\n",
       "    'type': 'string'},\n",
       "   'data_type': {'description': 'The nature or genre of the resource. Recommended best practice is to use a controlled vocabulary such as the DCMI Type Vocabulary [DCMITYPE]. To describe the file format, physical medium, or dimensions of the resource, use the Format element.\\n',\n",
       "    'type': 'string'},\n",
       "   'projects': {'anyOf': [{'type': 'array',\n",
       "      'items': {'type': 'object',\n",
       "       'additionalProperties': True,\n",
       "       'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "          'termDef': {'term': 'Universally Unique Identifier',\n",
       "           'source': 'NCIt',\n",
       "           'cde_id': 'C54100',\n",
       "           'cde_version': None,\n",
       "           'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "         'type': 'string',\n",
       "         'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "        'code': {'type': 'string'}},\n",
       "       'minItems': 1,\n",
       "       'maxItems': 1}},\n",
       "     {'type': 'object',\n",
       "      'additionalProperties': True,\n",
       "      'properties': {'id': {'term': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "         'termDef': {'term': 'Universally Unique Identifier',\n",
       "          'source': 'NCIt',\n",
       "          'cde_id': 'C54100',\n",
       "          'cde_version': None,\n",
       "          'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "        'type': 'string',\n",
       "        'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "       'code': {'type': 'string'}}}]}}}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "schema_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "91a11348",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-09 10:46:43,128 [INFO] Successfully loaded JSON file: ../examples/schema/json/schema_dev.json\n",
      "2025-10-09 10:46:43,131 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/demographic.yaml\n",
      "2025-10-09 10:46:43,136 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/project.yaml\n",
      "2025-10-09 10:46:43,137 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/acknowledgement.yaml\n",
      "2025-10-09 10:46:43,139 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/medical_history.yaml\n",
      "2025-10-09 10:46:43,146 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/_definitions.yaml\n",
      "2025-10-09 10:46:43,146 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/_settings.yaml\n",
      "2025-10-09 10:46:43,148 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/program.yaml\n",
      "2025-10-09 10:46:43,149 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/sample.yaml\n",
      "2025-10-09 10:46:43,192 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/_terms.yaml\n",
      "2025-10-09 10:46:43,194 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/lipidomics_file.yaml\n",
      "2025-10-09 10:46:43,195 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/publication.yaml\n",
      "2025-10-09 10:46:43,196 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/subject.yaml\n",
      "2025-10-09 10:46:43,199 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/core_metadata_collection.yaml\n",
      "2025-10-09 10:46:43,199 [INFO] Initializing DataDictionary with schema path: ../examples/schema/json/schema_dev.json\n",
      "2025-10-09 10:46:43,200 [INFO] Initializing ResolveSchema with schema path: ../examples/schema/json/schema_dev.json\n",
      "2025-10-09 10:46:43,200 [INFO] Starting schema resolution process.\n",
      "2025-10-09 10:46:43,200 [INFO] Reading JSON file from path: ../examples/schema/json/schema_dev.json\n",
      "2025-10-09 10:46:43,201 [INFO] Successfully read JSON schema.\n",
      "2025-10-09 10:46:43,201 [INFO] Retrieving node names from schema.\n",
      "2025-10-09 10:46:43,201 [INFO] Splitting schema into individual node schemas.\n",
      "2025-10-09 10:46:43,201 [INFO] Split schema into individual node schemas.\n",
      "2025-10-09 10:46:43,201 [INFO] Retrieving schema for schema ID: _definitions.yaml\n",
      "2025-10-09 10:46:43,201 [INFO] Retrieved definitions schema.\n",
      "2025-10-09 10:46:43,201 [INFO] Retrieving schema for schema ID: _terms.yaml\n",
      "2025-10-09 10:46:43,202 [INFO] Retrieved terms schema.\n",
      "2025-10-09 10:46:43,202 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,202 [INFO] Resolved references in definitions schema.\n",
      "2025-10-09 10:46:43,202 [INFO] Resolving all references in schema list.\n",
      "2025-10-09 10:46:43,202 [INFO] === Resolving Schema References ===\n",
      "2025-10-09 10:46:43,203 [INFO] Retrieving node names from schema.\n",
      "2025-10-09 10:46:43,203 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,203 [INFO] Resolved demographic.yaml\n",
      "2025-10-09 10:46:43,203 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,203 [INFO] Resolved project.yaml\n",
      "2025-10-09 10:46:43,203 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,204 [INFO] Resolved acknowledgement.yaml\n",
      "2025-10-09 10:46:43,204 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,204 [INFO] Resolved medical_history.yaml\n",
      "2025-10-09 10:46:43,204 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,204 [INFO] Resolved _settings.yaml\n",
      "2025-10-09 10:46:43,205 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,205 [INFO] Resolved program.yaml\n",
      "2025-10-09 10:46:43,205 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,205 [INFO] Resolved sample.yaml\n",
      "2025-10-09 10:46:43,205 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,205 [INFO] Resolved lipidomics_file.yaml\n",
      "2025-10-09 10:46:43,205 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,206 [INFO] Resolved publication.yaml\n",
      "2025-10-09 10:46:43,206 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,206 [INFO] Resolved subject.yaml\n",
      "2025-10-09 10:46:43,206 [INFO] Resolving references in schema.\n",
      "2025-10-09 10:46:43,206 [INFO] Resolved core_metadata_collection.yaml\n",
      "2025-10-09 10:46:43,206 [INFO] Resolved all references in schema list.\n",
      "2025-10-09 10:46:43,207 [INFO] Converting schema list to JSON format.\n",
      "2025-10-09 10:46:43,207 [INFO] Converted resolved schema list to JSON format.\n",
      "2025-10-09 10:46:43,210 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/demographic.yaml\n",
      "2025-10-09 10:46:43,214 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/project.yaml\n",
      "2025-10-09 10:46:43,218 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/acknowledgement.yaml\n",
      "2025-10-09 10:46:43,222 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/medical_history.yaml\n",
      "2025-10-09 10:46:43,223 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/program.yaml\n",
      "2025-10-09 10:46:43,227 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/sample.yaml\n",
      "2025-10-09 10:46:43,233 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/lipidomics_file.yaml\n",
      "2025-10-09 10:46:43,237 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/publication.yaml\n",
      "2025-10-09 10:46:43,240 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/subject.yaml\n",
      "2025-10-09 10:46:43,246 [INFO] Successfully wrote YAML file: ../tests/gen3_schema/examples/yaml/resolved/core_metadata_collection.yaml\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "demographic.yaml\n",
      "project.yaml\n",
      "acknowledgement.yaml\n",
      "medical_history.yaml\n",
      "program.yaml\n",
      "sample.yaml\n",
      "lipidomics_file.yaml\n",
      "publication.yaml\n",
      "subject.yaml\n",
      "core_metadata_collection.yaml\n"
     ]
    }
   ],
   "source": [
    "from gen3_validator.resolve_schema import ResolveSchema\n",
    "from gen3schemadev.utils import *\n",
    "\n",
    "# Generate a bundled json, a resolved bundled json, a dir of yamls and a dir of resolved yamls from a singled unresolved bundled_json\n",
    "\n",
    "bundled_schema_path = '../examples/schema/json/schema_dev.json'\n",
    "bundled_schema = read_json(bundled_schema_path)\n",
    "\n",
    "# write to yamls\n",
    "output_folder = \"../tests/gen3_schema/examples/yaml/\"\n",
    "for k,v in bundled_schema.items():\n",
    "    w_path = os.path.join(output_folder, k)\n",
    "    write_yaml(v, w_path)\n",
    "\n",
    "# Resolve the schema and write to yamls\n",
    "resolved_schema = resolve_schema(schema_path=bundled_schema_path)\n",
    "for v in resolved_schema:\n",
    "    k = v['id'] + '.yaml'\n",
    "    print(k)\n",
    "    w_path = os.path.join(output_folder, 'resolved', k)\n",
    "    write_yaml(v, w_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2368dec9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'demographic.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'demographic',\n",
       "  'title': 'Demographic',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Data for the characterization of the patient by means of segementing the population (e.g. characterization by age, sex, or race).',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'demographics',\n",
       "    'label': 'describes',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'year_birth': {'description': 'Year of birth at baseline (YYYY)',\n",
       "    'type': 'integer'},\n",
       "   'month_birth': {'description': 'Month of birth at baseline (MM)',\n",
       "    'type': 'integer'}}},\n",
       " 'project.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'project',\n",
       "  'title': 'Project',\n",
       "  'type': 'object',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'category': 'administrative',\n",
       "  'description': 'The study the data is coming from',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'state',\n",
       "   'released',\n",
       "   'releasable',\n",
       "   'intended_release_date'],\n",
       "  'required': ['code', 'name', 'programs', 'dbgap_accession_number'],\n",
       "  'uniqueKeys': [['id'], ['code']],\n",
       "  'links': [{'name': 'programs',\n",
       "    'backref': 'projects',\n",
       "    'label': 'member_of',\n",
       "    'target_type': 'program',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'constraints': None,\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'$ref': '_definitions.yaml#/UUID',\n",
       "    'systemAlias': 'node_id',\n",
       "    'description': 'UUID for the project.'},\n",
       "   'name': {'type': 'string',\n",
       "    'description': 'Display name/brief description for the project.'},\n",
       "   'code': {'type': 'string',\n",
       "    'description': 'Unique identifier for the project.'},\n",
       "   'date_collected': {'description': 'The date or date range in which the project data was collected.',\n",
       "    'type': 'string'},\n",
       "   'availability_type': {'description': 'Is the project open or restricted?',\n",
       "    'enum': ['Open', 'Restricted']},\n",
       "   'availability_mechanism': {'description': 'Mechanism by which the project will be made avilable.',\n",
       "    'type': 'string'},\n",
       "   'support_source': {'description': 'The name of source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'support_id': {'description': 'The ID of the source providing support/grant resources.',\n",
       "    'type': 'string'},\n",
       "   'programs': {'$ref': '_definitions.yaml#/to_one',\n",
       "    'description': 'Indicates that the Study is logically part of the indicated Study.\\n'},\n",
       "   'state': {'description': 'The possible states a project can be in.  All but `open` are\\nequivalent to some type of locked state.\\n',\n",
       "    'default': 'open',\n",
       "    'enum': ['open', 'review', 'submitted', 'processing', 'closed', 'legacy']},\n",
       "   'released': {'description': 'To release a project is to tell the GDC to include all submitted\\nnodes in the next GDC index.\\n',\n",
       "    'default': False,\n",
       "    'type': 'boolean'},\n",
       "   'releasable': {'description': 'A project can only be released by the user when `releasable` is true.\\n',\n",
       "    'default': False,\n",
       "    'type': 'boolean'},\n",
       "   'intended_release_date': {'description': \"Tracks a Project's intended release date.\",\n",
       "    'type': 'string',\n",
       "    'format': 'date-time'},\n",
       "   'dbgap_accession_number': {'type': 'string',\n",
       "    'description': 'The dbgap accession number provided for the project. (CMG)'},\n",
       "   'authz': {'description': 'The authz string (example, [\"/programs/Program.name/projects/Project.code\"] ).',\n",
       "    'type': 'array',\n",
       "    'items': {'type': 'string'}},\n",
       "   'short_name': {'description': 'The shorthand name for the project.',\n",
       "    'type': 'string'},\n",
       "   'full_name': {'description': 'The full name for the project.',\n",
       "    'type': 'string'},\n",
       "   'project_description': {'description': 'The description for the project.',\n",
       "    'type': 'string'},\n",
       "   'project_registration': {'description': 'External source from which the identifier included in study_id originates',\n",
       "    'type': 'string'},\n",
       "   'consent_codes': {'description': 'Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. \\\\n\\n        Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html',\n",
       "    'type': 'array'},\n",
       "   'data_access_url': {'description': 'A URL link that takes you to a page that explains how to gain access to the project',\n",
       "    'type': 'string'}},\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/'},\n",
       " 'acknowledgement.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'acknowledgement',\n",
       "  'title': 'Acknowledgement',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Acknowledgement of an individual or group involved in a project.',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'acknowledgements',\n",
       "    'label': 'contribute_to',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_many_project'}}},\n",
       " 'medical_history.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'medical_history',\n",
       "  'title': 'Medical History',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Medical history of the participant',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'medical_histories',\n",
       "    'label': 'describes',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'one_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'diabetes_type': {'description': 'diabetes diagnosed by fasting blood glucose >=7 mmol/l or tx AHA or 2 hour blood glucose >=11.1 mmol/l.',\n",
       "    'termDef': [{'term': 'Diabetes mellitus',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0000819',\n",
       "      'term_version': '2021-10-10'}],\n",
       "    'enum': ['IGT', 'KDM', 'IFG', 'NDM', 'NGT'],\n",
       "    'enumDef': [{'enumeration': 'IGT',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0040270'},\n",
       "     {'enumeration': 'NDM', 'source': 'SNOMED', 'term_id': '870528001'},\n",
       "     {'enumeration': 'NGT', 'source': 'SNOMED', 'term_id': '166926006'}]},\n",
       "   'hypertension': {'description': 'Whether the participant has Hypertension',\n",
       "    'enum': ['yes, measured or on treatment',\n",
       "     'yes, self-reported',\n",
       "     'no',\n",
       "     'not reported',\n",
       "     'not collected']},\n",
       "   'hypertension_measurement_type': {'description': 'Whether the hypertension was measured at recruitment/patients on hypertension treatment or self-reported medical diagnosis',\n",
       "    'enum': ['self-reported', 'measured']},\n",
       "   'diabetes': {'description': 'Self-reported diabetes in the participant',\n",
       "    'termDef': [{'term': 'Diabetes mellitus',\n",
       "      'source': 'hpo',\n",
       "      'term_id': 'HP:0000819'}],\n",
       "    'enum': ['yes', 'no', 'unsure', 'not reported', 'not collected']}}},\n",
       " '_definitions.yaml': {'id': '_definitions',\n",
       "  'UUID': {'term': {'$ref': '_terms.yaml#/UUID'},\n",
       "   'type': 'string',\n",
       "   'pattern': '^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$'},\n",
       "  'parent_uuids': {'type': 'array',\n",
       "   'minItems': 1,\n",
       "   'items': {'$ref': '#/UUID'},\n",
       "   'uniqueItems': True},\n",
       "  'foreign_key_project': {'type': 'object',\n",
       "   'additionalProperties': True,\n",
       "   'properties': {'id': {'$ref': '#/UUID'}, 'code': {'type': 'string'}}},\n",
       "  'to_one_project': {'anyOf': [{'type': 'array',\n",
       "     'items': {'$ref': '#/foreign_key_project', 'minItems': 1, 'maxItems': 1}},\n",
       "    {'$ref': '#/foreign_key_project'}]},\n",
       "  'to_many_project': {'anyOf': [{'type': 'array',\n",
       "     'items': {'$ref': '#/foreign_key_project', 'minItems': 1}},\n",
       "    {'$ref': '#/foreign_key_project'}]},\n",
       "  'foreign_key': {'type': 'object',\n",
       "   'additionalProperties': True,\n",
       "   'properties': {'id': {'$ref': '#/UUID'},\n",
       "    'submitter_id': {'type': 'string'}}},\n",
       "  'to_one': {'anyOf': [{'type': 'array',\n",
       "     'items': {'$ref': '#/foreign_key', 'minItems': 1, 'maxItems': 1}},\n",
       "    {'$ref': '#/foreign_key'}]},\n",
       "  'to_many': {'anyOf': [{'type': 'array',\n",
       "     'items': {'$ref': '#/foreign_key', 'minItems': 1}},\n",
       "    {'$ref': '#/foreign_key'}]},\n",
       "  'datetime': {'oneOf': [{'type': 'string', 'format': 'date-time'},\n",
       "    {'type': 'null'}],\n",
       "   'term': {'$ref': '_terms.yaml#/datetime'}},\n",
       "  'file_name': {'type': 'string', 'term': {'$ref': '_terms.yaml#/file_name'}},\n",
       "  'file_size': {'type': 'integer', 'term': {'$ref': '_terms.yaml#/file_size'}},\n",
       "  'file_format': {'type': 'string',\n",
       "   'term': {'$ref': '_terms.yaml#/file_format'}},\n",
       "  'ga4gh_drs_uri': {'type': 'string',\n",
       "   'term': {'$ref': '_terms.yaml#/ga4gh_drs_uri'}},\n",
       "  'md5sum': {'type': 'string',\n",
       "   'term': {'$ref': '_terms.yaml#/md5sum'},\n",
       "   'pattern': '^[a-f0-9]{32}$'},\n",
       "  'object_id': {'type': 'string',\n",
       "   'description': 'The GUID of the object in the index service.'},\n",
       "  'release_state': {'description': 'Release state of an node.',\n",
       "   'default': 'unreleased',\n",
       "   'enum': ['unreleased', 'released', 'redacted']},\n",
       "  'data_bundle_state': {'description': 'State of a data bundle.',\n",
       "   'default': 'submitted',\n",
       "   'enum': ['submitted',\n",
       "    'validated',\n",
       "    'error',\n",
       "    'released',\n",
       "    'suppressed',\n",
       "    'redacted']},\n",
       "  'data_file_error_type': {'term': {'$ref': '_terms.yaml#/data_file_error_type'},\n",
       "   'enum': ['file_size', 'file_format', 'md5sum']},\n",
       "  'state': {'term': {'$ref': '_terms.yaml#/state'},\n",
       "   'default': 'validated',\n",
       "   'downloadable': ['uploaded',\n",
       "    'md5summed',\n",
       "    'validating',\n",
       "    'validated',\n",
       "    'error',\n",
       "    'invalid',\n",
       "    'released'],\n",
       "   'public': ['live'],\n",
       "   'oneOf': [{'enum': ['uploading',\n",
       "      'uploaded',\n",
       "      'md5summing',\n",
       "      'md5summed',\n",
       "      'validating',\n",
       "      'error',\n",
       "      'invalid',\n",
       "      'suppressed',\n",
       "      'redacted',\n",
       "      'live']},\n",
       "    {'enum': ['validated', 'submitted', 'released']}]},\n",
       "  'file_state': {'term': {'$ref': '_terms.yaml#/file_state'},\n",
       "   'default': 'registered',\n",
       "   'enum': ['registered',\n",
       "    'uploading',\n",
       "    'uploaded',\n",
       "    'validating',\n",
       "    'validated',\n",
       "    'submitted',\n",
       "    'processing',\n",
       "    'processed',\n",
       "    'released',\n",
       "    'error']},\n",
       "  'qc_metrics_state': {'term': {'$ref': '_terms.yaml#/qc_metric_state'},\n",
       "   'enum': ['FAIL', 'PASS', 'WARN']},\n",
       "  'project_id': {'type': 'string',\n",
       "   'term': {'$ref': '_terms.yaml#/project_id'}},\n",
       "  'data_file_properties': {'$ref': '#/ubiquitous_properties',\n",
       "   'file_name': {'$ref': '#/file_name'},\n",
       "   'file_size': {'$ref': '#/file_size'},\n",
       "   'file_format': {'$ref': '#/file_format'},\n",
       "   'md5sum': {'$ref': '#/md5sum'},\n",
       "   'object_id': {'$ref': '#/object_id'},\n",
       "   'file_state': {'$ref': '#/file_state'},\n",
       "   'error_type': {'$ref': '#/data_file_error_type'},\n",
       "   'ga4gh_drs_uri': {'$ref': '#/ga4gh_drs_uri'}},\n",
       "  'workflow_properties': {'$ref': '#/ubiquitous_properties',\n",
       "   'workflow_link': {'description': 'Link to Github hash for the CWL workflow used.',\n",
       "    'type': 'string'},\n",
       "   'workflow_version': {'description': 'Major version for a GDC workflow.',\n",
       "    'type': 'string'},\n",
       "   'workflow_start_datetime': {'$ref': '#/datetime'},\n",
       "   'workflow_end_datetime': {'$ref': '#/datetime'}},\n",
       "  'ubiquitous_properties': {'type': {'type': 'string'},\n",
       "   'id': {'$ref': '#/UUID', 'systemAlias': 'node_id'},\n",
       "   'submitter_id': {'type': 'string',\n",
       "    'description': 'A project-specific identifier for a node. This property is the calling card/nickname/alias for a unit of submission. It can be used in place of the UUID for identifying or recalling a node.\\n'},\n",
       "   'state': {'$ref': '#/state'},\n",
       "   'project_id': {'$ref': '#/project_id'},\n",
       "   'created_datetime': {'$ref': '#/datetime'},\n",
       "   'updated_datetime': {'$ref': '#/datetime'}}},\n",
       " '_settings.yaml': {'enable_case_cache': False,\n",
       "  '_dict_commit': '520a25999fd183f6c5b7ddef2980f3e839517da5',\n",
       "  '_dict_version': '0.2.1-9-g520a259'},\n",
       " 'program.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'program',\n",
       "  'title': 'Program',\n",
       "  'type': 'object',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'A broad framework of goals to be achieved. (NCIt C52647)\\n',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': False,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id'],\n",
       "  'required': ['name', 'dbgap_accession_number'],\n",
       "  'uniqueKeys': [['id'], ['name']],\n",
       "  'links': [],\n",
       "  'properties': {'type': {'type': 'string'},\n",
       "   'id': {'$ref': '_definitions.yaml#/UUID', 'systemAlias': 'node_id'},\n",
       "   'name': {'type': 'string',\n",
       "    'description': 'Full name/title of the program.'},\n",
       "   'dbgap_accession_number': {'type': 'string',\n",
       "    'description': 'The dbgap accession number provided for the program.'}}},\n",
       " 'sample.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'sample',\n",
       "  'title': 'Sample',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'biospecimen',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': \"Biospecimen information that links subjects to samples including sample's provider and source.\",\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'subjects',\n",
       "    'backref': 'samples',\n",
       "    'label': 'taken_from',\n",
       "    'target_type': 'subject',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'subjects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'subjects': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'sample_provider': {'description': 'The name of collaborating institute that provided the sample',\n",
       "    'enum': ['Baker', 'USYD', 'UMELB', 'UQ']},\n",
       "   'sample_source': {'description': 'Uberon identifier, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG, CCDG)',\n",
       "    'type': 'string',\n",
       "    'pattern': '^UBERON:[0-9]{7}$'},\n",
       "   'sample_type': {'description': 'A string saying the type of sample',\n",
       "    'type': 'string'}}},\n",
       " '_terms.yaml': {'id': '_terms',\n",
       "  '28s_16s_ribosomal_rna_ratio': {'description': 'The 28S/18S ribosomal RNA band ratio used to assess the quality of total RNA.\\n',\n",
       "   'termDef': {'term': '28s/18s Ribosomal RNA Ratio',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'a260_a280_ratio': {'description': 'Numeric value that represents the sample ratio of nucleic acid absorbance at 260 nm and 280 nm, used to determine a measure of DNA purity.\\n',\n",
       "   'termDef': {'term': 'Nucleic Acid Absorbance at 260 And Absorbance at 280 DNA Purity Ratio Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432595,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432595&version=1.0'}},\n",
       "  'age': {'description': 'Age at last visit. Age is encoded as ISO8601 duration. (CMG)\\n',\n",
       "   'termDef': {'term': 'Age at last visit. Age is encoded as ISO8601 duration.',\n",
       "    'source': 'Phenopackets',\n",
       "    'cde_id': 'ISO8601',\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://phenopackets-schema.readthedocs.io/en/latest/age.html#rstage'}},\n",
       "  'family_onset': {'description': 'Typical age of onset for the family. Age is encoded as ISO8601 duration. (CMG)\\n',\n",
       "   'termDef': {'term': 'Typical age of onset for the family. Age is encoded as ISO8601 duration.',\n",
       "    'source': 'Phenopackets',\n",
       "    'cde_id': 'ISO8601',\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://phenopackets-schema.readthedocs.io/en/latest/age.html#rstage'}},\n",
       "  'disease_omim_id': {'description': 'The ID of the term from OMIM which represents a disease. (CMG)\\n',\n",
       "   'termDef': {'term': 'The ID of the term from OMIM which represents a disease. (CMG)',\n",
       "    'source': 'OMIM',\n",
       "    'cde_id': 'OMIM',\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://www.ncbi.nlm.nih.gov/omim'}},\n",
       "  'phenotype_hpo_id': {'description': 'The ID of the term from Human Phenotype Ontology which represents a harmonized phenotype. (CMG)\\n',\n",
       "   'termDef': {'term': 'The ID of the term from Human Phenotype Ontology which represents a harmonized phenotype. (CMG)',\n",
       "    'source': 'HPO',\n",
       "    'cde_id': 'HPO',\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://hpo.jax.org/app/'}},\n",
       "  'adapter_name': {'description': 'Name of the sequencing adapter.\\n'},\n",
       "  'adapter_sequence': {'description': 'Base sequence of the sequencing adapter.\\n'},\n",
       "  'age_of_onset': {'description': 'Age of onset (if affected). Otherwise age of last exam or phenotype transfer. (CMG)\\n',\n",
       "   'termDef': {'term': 'Age of onset (if affected). Otherwise age of last exam or phenotype transfer. (CMG)',\n",
       "    'source': 'Phenopackets',\n",
       "    'cde_id': 'ISO8601',\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://phenopackets-schema.readthedocs.io/en/latest/age.html#rstage'}},\n",
       "  'age_at_event_days_gt89': {'description': \"Indicate whether the 'Age at the time an event occurs expressed in the number of days since birth' is greater than 32872 days (89 years).\\n\"},\n",
       "  'age_at_diagnosis': {'description': \"Age at the time of diagnosis expressed in number of days since birth. If the age at diagnosis is greater than 32872 days (89 years), then see 'age_at_diagnosis_gt89'.\\n\",\n",
       "   'termDef': {'term': 'Patient Diagnosis Age Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3225640,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3225640&version=2.0'}},\n",
       "  'age_at_diagnosis_gt89': {'description': 'Indicate whether the age at the time of diagnosis expressed in number of days since birth is greater than 32872 days (89 years).\\n',\n",
       "   'termDef': {'term': 'Patient Diagnosis Age Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3225640,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3225640&version=2.0'}},\n",
       "  'ajcc_clinical_m': {'description': 'Extent of the distant metastasis for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.\\n',\n",
       "   'termDef': {'term': 'Neoplasm American Joint Committee on Cancer Clinical Distant Metastasis M Stage',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3440331,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440331&version=1.0'}},\n",
       "  'ajcc_clinical_n': {'description': 'Extent of the regional lymph node involvement for the cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.\\n',\n",
       "   'termDef': {'term': 'Neoplasm American Joint Committee on Cancer Clinical Regional Lymph Node N Stage',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3440330,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440330&version=1.0'}},\n",
       "  'ajcc_clinical_stage': {'description': 'Stage group determined from clinical information on the tumor (T), regional node (N) and metastases (M) and by grouping cases with similar prognosis for cancer.\\n',\n",
       "   'termDef': {'term': 'Neoplasm American Joint Committee on Cancer Clinical Group Stage',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3440332,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440332&version=1.0'}},\n",
       "  'ajcc_clinical_t': {'description': 'Extent of the primary cancer based on evidence obtained from clinical assessment parameters determined prior to treatment.\\n',\n",
       "   'termDef': {'term': 'Neoplasm American Joint Committee on Cancer Clinical Primary Tumor T Stage',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3440328,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3440328&version=1.0'}},\n",
       "  'ajcc_pathologic_m': {'description': 'Code to represent the defined absence or presence of distant spread or metastases (M) to locations via vascular channels or lymphatics beyond the regional lymph nodes, using criteria established by the American Joint Committee on Cancer (AJCC).\\n',\n",
       "   'termDef': {'term': 'American Joint Committee on Cancer Metastasis Stage Code',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3045439,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3045439&version=1.0'}},\n",
       "  'ajcc_pathologic_n': {'description': \"The codes that represent the stage of cancer based on the nodes present (N stage) according to criteria based on multiple editions of the AJCC's Cancer Staging Manual.\\n\",\n",
       "   'termDef': {'term': 'Neoplasm Disease Lymph Node Stage American Joint Committee on Cancer Code',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3203106,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3203106&version=1.0'}},\n",
       "  'ajcc_pathologic_stage': {'description': 'The extent of a cancer, especially whether the disease has spread from the original site to other parts of the body based on AJCC staging criteria.\\n',\n",
       "   'termDef': {'term': 'Neoplasm Disease Stage American Joint Committee on Cancer Code',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3203222,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3203222&version=1.0'}},\n",
       "  'ajcc_pathologic_t': {'description': 'Code of pathological T (primary tumor) to define the size or contiguous extension of the primary tumor (T), using staging criteria from the American Joint Committee on Cancer (AJCC).\\n',\n",
       "   'termDef': {'term': 'American Joint Committee on Cancer Tumor Stage Code',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3045435,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3045435&version=1.0'}},\n",
       "  'alcohol_history': {'description': 'A response to a question that asks whether the participant has consumed at least 12 drinks of any kind of alcoholic beverage in their lifetime.\\n',\n",
       "   'termDef': {'term': 'Alcohol Lifetime History Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2201918,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2201918&version=1.0'}},\n",
       "  'drink_number': {'description': 'Units of drink consumed during Drink Period (DRNKPRD). (GTEx)\\n'},\n",
       "  'aliquot_quantity': {'description': 'The quantity in micrograms (ug) of the aliquot(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Aliquot Quantity',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'aliquot_volume': {'description': 'The volume in microliters (ml) of the aliquot(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Aliquot Volume',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'amount': {'description': 'Weight in grams or volume in mL.\\n'},\n",
       "  'analyte_quantity': {'description': 'The quantity in micrograms (ug) of the analyte(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Analyte Quantity',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'analyte_type': {'description': 'Text term that represents the kind of molecular specimen analyte.\\n',\n",
       "   'termDef': {'term': 'Molecular Specimen Type Text Name',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2513915,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2513915&version=2.0'}},\n",
       "  'analyte_type_id': {'description': 'A single letter code used to identify a type of molecular analyte.\\n',\n",
       "   'termDef': {'term': 'Molecular Analyte Identification Code',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432508,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432508&version=1.0'}},\n",
       "  'analyte_volume': {'description': 'The volume in microliters (ml) of the analyte(s) derived from the analyte(s) shipped for sequencing and characterization.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Analyte Volume',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'ann_arbor_b_symptoms': {'description': \"Text term to signify whether lymphoma B-symptoms are present as noted in the patient's medical record.\\n\",\n",
       "   'termDef': {'term': 'Lymphoma B-Symptoms Medical Record Documented Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2902402,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2902402&version=1.0'}},\n",
       "  'ann_arbor_clinical_stage': {'description': \"The classification of the clinically confirmed anatomic disease extent of lymphoma (Hodgkin's and Non-Hodgkins) based on the Ann Arbor Staging System.\\n\",\n",
       "   'termDef': {'term': 'Ann Arbor Clinical Stage',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'ann_arbor_extranodal_involvement': {'description': 'Indicator that identifies whether a patient with malignant lymphoma has lymphomatous involvement of an extranodal site.\\n',\n",
       "   'termDef': {'term': 'Lymphomatous Extranodal Site Involvement Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3364582,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3364582&version=1.0'}},\n",
       "  'ann_arbor_pathologic_stage': {'description': \"The classification of the pathologically confirmed anatomic disease extent of lymphoma (Hodgkin's and Non-Hodgkins) based on the Ann Arbor Staging System.\\n\",\n",
       "   'termDef': {'term': 'Ann Arbor Pathologic Stage',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'ann_arbor_tumor_stage': {'description': \"The classification of the anatomic disease extent of lymphoma (Hodgkin's and Non-Hodgkins) based on the Ann Arbor Staging System.\\n\",\n",
       "   'termDef': {'term': 'Ann Arbor Tumor Stage',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'base_caller_name': {'description': 'Name of the base caller.\\n'},\n",
       "  'base_caller_version': {'description': 'Version of the base caller.\\n'},\n",
       "  'biomarker_name': {'description': 'The name of the biomarker being tested for this specimen and set of test results.\\n',\n",
       "   'termDef': {'term': 'Biomarker Name',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5473,\n",
       "    'cde_version': 11,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5473&version=2.31'}},\n",
       "  'biomarker_result': {'description': 'Text term to define the results of genetic testing.\\n',\n",
       "   'termDef': {'term': 'Laboratory Procedure Genetic Abnormality Test Result Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3234680,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3234680&version=1.0'}},\n",
       "  'biomarker_test_method': {'description': 'Text descriptor of a molecular analysis method used for an individual.\\n',\n",
       "   'termDef': {'term': 'Disease Detection Molecular Analysis Method Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3121575,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3121575&version=1.0'}},\n",
       "  'biospecimen_anatomic_site_id': {'description': 'Uberon ID, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)\\n',\n",
       "   'termDef': {'term': 'Uberon ID, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)',\n",
       "    'source': 'Uberon',\n",
       "    'cde_id': 'Uberon',\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://www.ebi.ac.uk/ols/ontologies/uberon'}},\n",
       "  'biospecimen_anatomic_site_label': {'description': 'Uberon Label, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)\\n',\n",
       "   'termDef': {'term': 'Uberon Label, anatomical location as described by the Uber Anatomy Ontology (UBERON). (CMG)',\n",
       "    'source': 'Uberon',\n",
       "    'cde_id': 'Uberon',\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://www.ebi.ac.uk/ols/ontologies/uberon'}},\n",
       "  'bmi': {'description': 'Autocalculated field of BMI: general indicator of the body fat an individual is carrying based upon the ratio of weight to height. BMI = 703 * weight in pounds/height in inches squared. (GTEx)\\n',\n",
       "   'termDef': {'term': 'Body Mass Index (BMI)',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 4973892,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4973892&version=1.0'}},\n",
       "  'burkitt_lymphoma_clinical_variant': {'description': \"Burkitt's lymphoma categorization based on clinical features that differ from other forms of the same disease.\\n\",\n",
       "   'termDef': {'term': 'Burkitt Lymphoma Clinical Variant Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3770421,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3770421&version=1.0'}},\n",
       "  'cause_of_death': {'description': 'Text term to identify the cause of death for a patient.\\n',\n",
       "   'termDef': {'term': 'Patient Death Reason',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2554674,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2554674&version=3.0'}},\n",
       "  'cea_level_preoperative': {'description': 'Numeric value of the Carcinoembryonic antigen or CEA at the time before surgery. [Manually- curated]\\n',\n",
       "   'termDef': {'term': 'Preoperative Carcinoembryonic Antigen Result Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2716510,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2716510&version=1.0'}},\n",
       "  'cigarettes_per_day': {'description': 'The average number of cigarettes smoked per day.\\n',\n",
       "   'termDef': {'term': 'Smoking Use Average Number',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2001716,\n",
       "    'cde_version': 4,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2001716&version=4.0'}},\n",
       "  'circumferential_resection_margin': {'description': 'A value in millimeters indicating the measured length between a malignant lesion of the colon or rectum and the nearest radial (or circumferential) border of tissue removed during cancer surgery.\\n',\n",
       "   'termDef': {'term': 'Colorectal Surgical Margin Circumferential Distance Measurement',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 64202,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64202&version=3.0'}},\n",
       "  'classification_of_tumor': {'description': 'Text that describes the kind of disease present in the tumor specimen as related to a specific timepoint.\\n',\n",
       "   'termDef': {'term': 'Tumor Tissue Disease Description Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3288124,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3288124&version=1.0'}},\n",
       "  'colon_polyps_history': {'description': 'Yes/No indicator to describe if the subject had a previous history of colon polyps as noted in the history/physical or previous endoscopic report (s).\\n',\n",
       "   'termDef': {'term': 'Colon Carcinoma Polyp Occurrence Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3107197,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3107197&version=1.0'}},\n",
       "  'composition': {'description': 'Text term that represents the cellular composition of the sample.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Cellular Composition Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432591,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432591&version=1.0'}},\n",
       "  'concentration': {'description': 'Numeric value that represents the concentration of an analyte or aliquot extracted from the sample or sample portion, measured in milligrams per milliliter.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Analyte or Aliquot Extracted Concentration Milligram per Milliliter Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432594,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432594&version=1.0'}},\n",
       "  'creation_datetime': {'description': 'The datetime of portion creation encoded as seconds from epoch.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Portion Creation Seconds Date/Time',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432592,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432592&version=1.0'}},\n",
       "  'current_weight': {'description': 'Numeric value that represents the current weight of the sample, measured  in milligrams.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Current Weight Milligram Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432606,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432606&version=1.0'}},\n",
       "  'data_category': {'description': 'Broad categorization of the contents of the data file.\\n'},\n",
       "  'data_file_error_type': {'description': 'Type of error for the data file object.\\n'},\n",
       "  'data_format': {'description': 'Format of the data files.\\n'},\n",
       "  'data_type': {'description': 'Specific content type of the data file. (CMG)\\n'},\n",
       "  'datetime': {'description': 'A combination of date and time of day in the form [-]CCYY-MM-DDThh:mm:ss[Z|(+|-)hh:mm]\\n'},\n",
       "  'days_to_birth': {'description': \"Time interval from a person's date of birth to the date of initial pathologic diagnosis, represented as a calculated negative number of days.\\n\",\n",
       "   'termDef': {'term': 'Person Birth Date Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3008233,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008233&version=1.0'}},\n",
       "  'days_to_collection': {'description': 'Time interval from the date of biospecimen collection to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Collection Date Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3008340,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008340&version=1.0'}},\n",
       "  'days_to_death': {'description': \"Time interval from a person's date of death to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n\",\n",
       "   'termDef': {'term': 'Death Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3165475,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3165475&version=1.0'}},\n",
       "  'days_to_hiv_diagnosis': {'description': 'Time interval from the date of the initial pathologic diagnosis to the date of human immunodeficiency diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'Human Immunodeficiency Virus Diagnosis Subtract Initial Pathologic Diagnosis Time Duration Day Calculation Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 4618491,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4618491&version=1.0'}},\n",
       "  'days_to_last_follow_up': {'description': 'Time interval from the date of last follow up to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'Last Communication Contact Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3008273,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008273&version=1.0'}},\n",
       "  'days_to_last_known_disease_status': {'description': 'Time interval from the date of last follow up to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'Last Communication Contact Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3008273,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008273&version=1.0'}},\n",
       "  'days_to_new_event': {'description': 'Time interval from the date of new tumor event including progression, recurrence and new primary malignacies to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'New Tumor Event Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3392464,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3392464&version=1.0'}},\n",
       "  'days_to_recurrence': {'description': 'Time interval from the date of new tumor event including progression, recurrence and new primary malignancies to the date of initial pathologic diagnosis, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'New Tumor Event Less Initial Pathologic Diagnosis Date Calculated Day Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3392464,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3392464&version=1.0'}},\n",
       "  'days_to_sample_procurement': {'description': 'The number of days from the date the patient was diagnosed to the date of the procedure that produced the sample.\\n'},\n",
       "  'days_to_treatment': {'description': 'Number of days from date of initial pathologic diagnosis that treatment began.\\n',\n",
       "   'termDef': {'term': 'Days to Treatment Start',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'days_to_treatment_end': {'description': 'Time interval from the date of the initial pathologic diagnosis to the date of treatment end, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'Treatment End Subtract First Pathologic Diagnosis Day Calculation Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5102431,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102431&version=1.0'}},\n",
       "  'days_to_treatment_start': {'description': 'Time interval from the date of the initial pathologic diagnosis to the start of treatment, represented as a calculated number of days.\\n',\n",
       "   'termDef': {'term': 'Treatment Start Subtract First Pathologic Diagnosis Time Day Calculation Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5102411,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102411&version=1.0'}},\n",
       "  'diagnosis_pathologically_confirmed': {'description': 'The histologic description of tissue or cells confirmed by a pathology review of frozen or formalin fixed slide(s) completed after the diagnostic pathology review of the tumor sample used to extract analyte(s).\\n',\n",
       "   'termDef': {'term': 'Post-Diagnostic Pathology Review Confirmation',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'dlco_ref_predictive_percent': {'description': \"The value, as a percentage of predicted lung volume, measuring the amount of carbon monoxide detected in a patient's lungs.\\n\",\n",
       "   'termDef': {'term': 'Lung Carbon Monoxide Diffusing Capability Test Assessment Predictive Value Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2180255,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2180255&version=1.0'}},\n",
       "  'encoding': {'description': 'Version of ASCII encoding of quality values found in the file.\\n',\n",
       "   'termDef': {'term': 'Encoding',\n",
       "    'source': 'FastQC',\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/1%20Basic%20Statistics.html'}},\n",
       "  'estrogen_receptor_percent_positive_ihc': {'description': 'Classification to represent ER Positive results expressed as a percentage value.\\n',\n",
       "   'termDef': {'term': 'ER Level Cell Percentage Category',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3128341,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3128341&version=1.0'}},\n",
       "  'estrogen_receptor_result_ihc': {'description': 'Text term to represent the overall result of Estrogen Receptor (ER) testing.\\n',\n",
       "   'termDef': {'term': 'Breast Carcinoma Estrogen Receptor Status',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2957359,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2957359&version=2.0'}},\n",
       "  'ethnicity': {'description': \"An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino (HARMONIZED). The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau. (CMG, GTEx)\\n\",\n",
       "   'termDef': {'term': 'Ethnic Group Category Text',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2192217,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2192217&version=2.0'}},\n",
       "  'experiment_name': {'description': 'Submitter-defined name for the experiment.\\n'},\n",
       "  'experimental_strategy': {'description': 'The sequencing strategy used to generate the data file referred to by the ga4gh_drs_uri.\\n'},\n",
       "  'fastq_name': {'description': 'Names of FASTQs.\\n'},\n",
       "  'fev1_ref_post_bronch_percent': {'description': 'The percentage comparison to a normal value reference range of the volume of air that a patient can forcibly exhale from the lungs in one second post-bronchodilator.\\n',\n",
       "   'termDef': {'term': 'Post Bronchodilator Lung Forced Expiratory Volume 1 Test Lab Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3302948,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302948&version=1.0'}},\n",
       "  'fev1_ref_pre_bronch_percent': {'description': 'The percentage comparison to a normal value reference range of the volume of air that a patient can forcibly exhale from the lungs in one second pre-bronchodilator.\\n',\n",
       "   'termDef': {'term': 'Pre Bronchodilator Lung Forced Expiratory Volume 1 Test Lab Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3302947,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302947&version=1.0'}},\n",
       "  'fev1_fvc_post_bronch_percent': {'description': 'Percentage value to represent result of Forced Expiratory Volume in 1 second (FEV1) divided by the Forced Vital Capacity (FVC) post-bronchodilator.\\n',\n",
       "   'termDef': {'term': 'Post Bronchodilator FEV1/FVC Percent Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3302956,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302956&version=1.0'}},\n",
       "  'fev1_fvc_pre_bronch_percent': {'description': 'Percentage value to represent result of Forced Expiratory Volume in 1 second (FEV1) divided by the Forced Vital Capacity (FVC) pre-bronchodilator.\\n',\n",
       "   'termDef': {'term': 'Pre Bronchodilator FEV1/FVC Percent Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3302955,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3302955&version=1.0'}},\n",
       "  'figo_stage': {'description': 'The extent of a cervical or endometrial cancer within the body, especially whether the disease has spread from the original site to other parts of the body, as described by the International Federation of Gynecology and Obstetrics (FIGO) stages.\\n',\n",
       "   'termDef': {'term': 'Gynecologic Tumor Grouping Cervical Endometrial FIGO 2009 Stage',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3225684,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3225684&version=1.0'}},\n",
       "  'file_format': {'description': 'The format of the data file object.\\n'},\n",
       "  'file_name': {'description': 'The name (or part of a name) of a file (of any type).\\n'},\n",
       "  'file_size': {'description': 'The size of the data file (object) in bytes.\\n'},\n",
       "  'file_state': {'description': 'The current state of the data file object.\\n'},\n",
       "  'flow_cell_barcode': {'description': 'Flow Cell Barcode.\\n'},\n",
       "  'freezing_method': {'description': 'Text term that represents the method used for freezing the sample.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Freezing Method Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432607,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432607&version=1.0'}},\n",
       "  'ga4gh_drs_uri': {'description': 'DRS URI as defined by GA4GH DRS spec for pointers to file objects.\\n'},\n",
       "  'sex': {'description': 'Text designations that identify gender (HARMONIZED). Gender is described as the assemblage of properties that distinguish people on the basis of their societal roles. [Explanatory Comment 1: Identification of gender is based upon self-report and may come from a form, questionnaire, interview, etc.] (CMG)\\n',\n",
       "   'termDef': {'term': 'Person Gender Text Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2200604,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2200604&version=3.0'}},\n",
       "  'height': {'description': 'The height of Donor in inches as reported by Donor, family/next of kin, or medical record abstraction. (GTEx)\\n',\n",
       "   'termDef': {'term': 'Patient Height Measurement',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 649,\n",
       "    'cde_version': 4.1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=649&version=4.1'}},\n",
       "  'her2_erbb2_percent_positive_ihc': {'description': 'Classification to represent the number of positive HER2/ERBB2 cells in a specimen or sample.\\n',\n",
       "   'termDef': {'term': 'HER2 ERBB Positive Finding Cell Percentage Category',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3086980,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3086980&version=1.0'}},\n",
       "  'her2_erbb2_result_fish': {'description': 'the type of outcome for HER2 as determined by an in situ hybridization (ISH) assay.\\n',\n",
       "   'termDef': {'term': 'Laboratory Procedure HER2/neu in situ Hybridization Outcome Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2854089,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2854089&version=1.0'}},\n",
       "  'her2_erbb2_result_ihc': {'description': 'Text term to signify the result of the medical procedure that involves testing a sample of blood or tissue for HER2 by histochemical localization of immunoreactive substances using labeled antibodies as reagents.\\n',\n",
       "   'termDef': {'term': 'Laboratory Procedure HER2/neu Immunohistochemistry Receptor Status',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2957563,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2957563&version=2.0'}},\n",
       "  'hiv_positive': {'description': 'Text term to signify whether a physician has diagnosed HIV infection in a patient.\\n',\n",
       "   'termDef': {'term': 'Physician Diagnosed HIV Infection Personal Medical History Yes No Not Applicable Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 4030799,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4030799&version=1.0'}},\n",
       "  'hpv_positive_type': {'description': 'Text classification to represent the strain or type of human papillomavirus identified in an individual.\\n',\n",
       "   'termDef': {'term': 'Human Papillomavirus Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2922649,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2922649&version=1.0'}},\n",
       "  'hpv_status': {'description': 'The findings of the oncogenic HPV.\\n',\n",
       "   'termDef': {'term': 'Oncogenic Human Papillomavirus Result Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2230033,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2230033&version=1.0'}},\n",
       "  'includes_spike_ins': {'description': 'Spike-in included?\\n'},\n",
       "  'initial_weight': {'description': 'Numeric value that represents the initial weight of the sample, measured in milligrams.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Initial Weight Milligram Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432605,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432605&version=1.0'}},\n",
       "  'instrument_model': {'description': 'Numeric value that represents the sample dimension that is greater than the shortest dimension and less than the longest dimension, measured in millimeters.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Intermediate Dimension Millimeter Measurement',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432604,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432604&version=1.0'}},\n",
       "  'intermediate_dimension': {'description': 'Intermediate dimension of the sample, in millimeters.\\n'},\n",
       "  'is_ffpe': {'description': 'Indicator to signify whether or not the tissue sample was fixed in formalin and embedded in paraffin (FFPE).\\n',\n",
       "   'termDef': {'term': 'Specimen Processing Formalin Fixed Paraffin Embedded Tissue Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 4170557,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4170557&version=1.0'}},\n",
       "  'is_paired_end': {'description': 'Are the reads paired end?\\n'},\n",
       "  'is_proband': {'description': 'Boolean value to determine if the case is the proband.\\n'},\n",
       "  'last_known_disease_status': {'description': \"Text term that describes the last known state or condition of an individual's neoplasm.\\n\",\n",
       "   'termDef': {'term': 'Person Last Known Neoplasm Status',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5424231,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2759550&version=1.0'}},\n",
       "  'laterality': {'description': 'For tumors in paired organs, designates the side on which the cancer originates.\\n',\n",
       "   'termDef': {'term': 'Primary Tumor Laterality',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 827,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=827&version=3.0'}},\n",
       "  'ldh_level_at_diagnosis': {'description': 'The 2 decimal place numeric laboratory value measured, assigned or computed related to the assessment of lactate dehydrogenase in a specimen.\\n',\n",
       "   'termDef': {'term': 'Laboratory Procedure Lactate Dehydrogenase Result Integer::2 Decimal Place Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2798766,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2798766&version=1.0'}},\n",
       "  'ldh_normal_range_upper': {'description': 'The top value of the range of statistical characteristics that are supposed to represent accepted standard, non-pathological pattern for lactate dehydrogenase (units not specified).\\n',\n",
       "   'termDef': {'term': 'Laboratory Procedure Lactate Dehydrogenase Result Upper Limit of Normal Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2597015,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2597015&version=1.0'}},\n",
       "  'library_strategy': {'description': 'Library strategy.\\n'},\n",
       "  'library_selection': {'description': 'Library Selection Method\\n'},\n",
       "  'library_name': {'description': 'Name of the library.\\n'},\n",
       "  'library_preparation_kit_name': {'description': 'Name of Library Preparation Kit\\n'},\n",
       "  'library_preparation_kit_vendor': {'description': 'Vendor of Library Preparation Kit\\n'},\n",
       "  'library_preparation_kit_catalog_number': {'description': 'Catalog of Library Preparation Kit\\n'},\n",
       "  'library_preparation_kit_version': {'description': 'Version of Library Preparation Kit\\n'},\n",
       "  'library_strand': {'description': 'Library stranded-ness.\\n'},\n",
       "  'longest_dimension': {'description': 'Numeric value that represents the longest dimension of the sample, measured in millimeters.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Longest Dimension Millimeter Measurement',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432602,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432602&version=1.0'}},\n",
       "  'lymph_nodes_positive': {'description': 'The number of lymph nodes involved with disease as determined by pathologic examination.\\n',\n",
       "   'termDef': {'term': 'Lymph Node(s) Positive Number',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 89,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=89&version=3.0'}},\n",
       "  'lymphatic_invasion_present': {'description': 'A yes/no indicator to ask if small or thin-walled vessel invasion is present, indicating lymphatic involvement\\n',\n",
       "   'termDef': {'term': 'Lymphatic/Small vessel Invasion Ind',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 64171,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64171&version=3.0'}},\n",
       "  'method_of_diagnosis': {'description': \"The method used to initially the patient's diagnosis.\\n\",\n",
       "   'termDef': {'term': 'Method of Diagnosis',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'method_of_sample_procurement': {'description': 'The method used to procure the sample used to extract analyte(s).\\n',\n",
       "   'termDef': {'term': 'Method of Sample Procurement',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'md5sum': {'description': \"The 128-bit hash value expressed as a 32 digit hexadecimal number used as a file's digital fingerprint.\\n\"},\n",
       "  'microsatellite_instability_abnormal': {'description': 'The yes/no indicator to signify the status of a tumor for microsatellite instability.\\n',\n",
       "   'termDef': {'term': 'Microsatellite Instability Occurrence Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3123142,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3123142&version=1.0'}},\n",
       "  'morphology': {'description': 'The third edition of the International Classification of Diseases for Oncology, published in 2000 used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms. The study of the structure of the cells and their arrangement to constitute tissues and, finally, the association among these to form organs. In pathology, the microscopic process of identifying normal and abnormal morphologic characteristics in tissues, by employing various cytochemical and immunocytochemical stains. A system of numbered categories for representation of data.\\n',\n",
       "   'termDef': {'term': 'International Classification of Diseases for Oncology, Third Edition ICD-O-3 Histology Code',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3226275,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3226275&version=1.0'}},\n",
       "  'new_event_anatomic_site': {'description': 'Text term to specify the anatomic location of the return of tumor after treatment.\\n',\n",
       "   'termDef': {'term': 'New Neoplasm Event Occurrence Anatomic Site',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3108271,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3108271&version=2.0'}},\n",
       "  'new_event_type': {'description': 'Text term to identify a new tumor event.\\n',\n",
       "   'termDef': {'term': 'New Neoplasm Event Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3119721,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3119721&version=1.0'}},\n",
       "  'normal_tumor_genotype_snp_match': {'description': 'Text term that represents whether or not the genotype of the normal tumor matches or if the data is not available.\\n',\n",
       "   'termDef': {'term': 'Normal Tumor Genotype Match Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 4588156,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4588156&version=1.0'}},\n",
       "  'number_proliferating_cells': {'description': 'Numeric value that represents the count of proliferating cells determined during pathologic review of the sample slide(s).\\n',\n",
       "   'termDef': {'term': 'Pathology Review Slide Proliferating Cell Count',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432636,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432636&version=1.0'}},\n",
       "  'oct_embedded': {'description': 'Indicator of whether or not the sample was embedded in Optimal Cutting Temperature (OCT) compound.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Optimal Cutting Temperature Compound Embedding Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432538,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432538&version=1.0'}},\n",
       "  'years_smoked': {'description': \"Numeric value (or unknown) to represent the number of years a person has been smoking (HARMONIZED). If the number of years is greater than 89, see 'years_smoked_gt89'.\\n\",\n",
       "   'termDef': {'term': 'Person Smoking Duration Year Count',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3137957,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3137957&version=1.0'}},\n",
       "  'percent_eosinophil_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by eosinophils in a tumor sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Specimen Eosinophilia Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2897700,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897700&version=2.0'}},\n",
       "  'percent_gc_content': {'description': 'The overall %GC of all bases in all sequences.\\n',\n",
       "   'termDef': {'term': '%GC',\n",
       "    'source': 'FastQC',\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/1%20Basic%20Statistics.html'}},\n",
       "  'percent_granulocyte_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by granulocytes in a tumor sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Specimen Granulocyte Infiltration Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2897705,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897705&version=2.0'}},\n",
       "  'percent_inflam_infiltration': {'description': 'Numeric value to represent local response to cellular injury, marked by capillary dilatation, edema and leukocyte infiltration; clinically, inflammation is manifest by reddness, heat, pain, swelling and loss of function, with the need to heal damaged tissue.\\n',\n",
       "   'termDef': {'term': 'Specimen Inflammation Change Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2897695,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897695&version=1.0'}},\n",
       "  'percent_lymphocyte_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by lymphocytes in a solid tissue normal sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Specimen Lymphocyte Infiltration Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2897710,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897710&version=2.0'}},\n",
       "  'percent_monocyte_infiltration': {'description': 'Numeric value to represent the percentage of monocyte infiltration in a sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Specimen Monocyte Infiltration Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5455535,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5455535&version=1.0'}},\n",
       "  'percent_necrosis': {'description': 'Numeric value to represent the percentage of cell death in a malignant tumor sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Malignant Neoplasm Necrosis Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2841237,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841237&version=1.0'}},\n",
       "  'percent_neutrophil_infiltration': {'description': 'Numeric value to represent the percentage of infiltration by neutrophils in a tumor sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Malignant Neoplasm Neutrophil Infiltration Percentage Cell Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2841267,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841267&version=1.0'}},\n",
       "  'percent_normal_cells': {'description': 'Numeric value to represent the percentage of normal cell content in a malignant tumor sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Malignant Neoplasm Normal Cell Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2841233,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841233&version=1.0'}},\n",
       "  'percent_stromal_cells': {'description': 'Numeric value to represent the percentage of reactive cells that are present in a malignant tumor sample or specimen but are not malignant such as fibroblasts, vascular structures, etc.\\n',\n",
       "   'termDef': {'term': 'Malignant Neoplasm Stromal Cell Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2841241,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841241&version=1.0'}},\n",
       "  'percent_tumor_cells': {'description': 'Numeric value that represents the percentage of infiltration by granulocytes in a sample.\\n',\n",
       "   'termDef': {'term': 'Specimen Tumor Cell Percentage Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432686,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432686&version=1.0'}},\n",
       "  'percent_tumor_nuclei': {'description': 'Numeric value to represent the percentage of tumor nuclei in a malignant neoplasm sample or specimen.\\n',\n",
       "   'termDef': {'term': 'Malignant Neoplasm Neoplasm Nucleus Percentage Cell Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2841225,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2841225&version=1.0'}},\n",
       "  'perineural_invasion_present': {'description': 'a yes/no indicator to ask if perineural invasion or infiltration of tumor or cancer is present.\\n',\n",
       "   'termDef': {'term': 'Tumor Perineural Invasion Ind',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 64181,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64181&version=3.0'}},\n",
       "  'platform': {'description': 'Name of the platform used to obtain data.\\n'},\n",
       "  'portion_number': {'description': 'Numeric value that represents the sequential number assigned to a portion of the sample.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Portion Sequence Number',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432711,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432711&version=1.0'}},\n",
       "  'portion_weight': {'description': 'Numeric value that represents the sample portion weight, measured in milligrams.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Portion Weight Milligram Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432593,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432593&version=1.0'}},\n",
       "  'preservation_method': {'description': 'Text term that represents the method used to preserve the sample.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Preservation Method Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432521,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432521&version=1.0'}},\n",
       "  'prior_malignancy': {'description': \"Text term to describe the patient's history of prior cancer diagnosis and the spatial location of any previous cancer occurrence.\\n\",\n",
       "   'termDef': {'term': 'Prior Cancer Diagnosis Occurrence Description Text',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3382736,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3382736&version=2.0'}},\n",
       "  'prior_treatment': {'description': 'A yes/no/unknown/not applicable indicator related to the administration of therapeutic agents received before the body specimen was collected.\\n',\n",
       "   'termDef': {'term': 'Therapeutic Procedure Prior Specimen Collection Administered Yes No Unknown Not Applicable Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 4231463,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4231463&version=1.0'}},\n",
       "  'progesterone_receptor_percent_positive_ihc': {'description': 'Classification to represent Progesterone Receptor Positive results expressed as a percentage value.\\n',\n",
       "   'termDef': {'term': 'Progesterone Receptor Level Cell Percentage Category',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3128342,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3128342&version=1.0'}},\n",
       "  'progesterone_receptor_result_ihc': {'description': 'Text term to represent the overall result of Progresterone Receptor (PR) testing.\\n',\n",
       "   'termDef': {'term': 'Breast Carcinoma Progesterone Receptor Status',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2957357,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2957357&version=2.0'}},\n",
       "  'progression_or_recurrence': {'description': 'Yes/No/Unknown indicator to identify whether a patient has had a new tumor event after initial treatment.\\n',\n",
       "   'termDef': {'term': 'New Neoplasm Event Post Initial Therapy Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3121376,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3121376&version=1.0'}},\n",
       "  'project_id': {'description': 'Unique ID for any specific defined piece of work that is undertaken or attempted to meet a single requirement.\\n'},\n",
       "  'qc_metric_state': {'description': 'State classification given by FASTQC for the metric. Metric specific details about the states are available on their website.\\n',\n",
       "   'termDef': {'term': 'QC Metric State',\n",
       "    'source': 'FastQC',\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/'}},\n",
       "  'race': {'description': 'An arbitrary classification of a taxonomic group that is a division of a species (HARMONIZED). It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau. (GTEx)\\n',\n",
       "   'termDef': {'term': 'Race Category Text',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2192199,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2192199&version=1.0'}},\n",
       "  'read_length': {'description': 'The length of the reads.\\n'},\n",
       "  'read_group_name': {'description': 'The name of the read group.\\n'},\n",
       "  'relationship_age_at_diagnosis': {'description': \"The age (in years) when the patient's relative was first diagnosed.\\n\",\n",
       "   'termDef': {'term': 'Relative Diagnosis Age Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5300571,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5300571&version=1.0'}},\n",
       "  'relationship_to_proband': {'description': 'The subgroup that describes the state of connectedness between members of the unit of society organized around kinship ties.\\n',\n",
       "   'termDef': {'term': 'Family Member Relationship Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2690165,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2690165&version=2.0'}},\n",
       "  'relationship_type': {'description': 'The subgroup that describes the state of connectedness between members of the unit of society organized around kinship ties.\\n',\n",
       "   'termDef': {'term': 'Family Member Relationship Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2690165,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2690165&version=2.0'}},\n",
       "  'relative_with_cancer_history': {'description': \"Indicator to signify whether or not an individual's biological relative has been diagnosed with another type of cancer.\\n\",\n",
       "   'termDef': {'term': 'Other Cancer Biological Relative History Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3901752,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3901752&version=1.0'}},\n",
       "  'residual_disease': {'description': 'Text terms to describe the status of a tissue margin following surgical resection.\\n',\n",
       "   'termDef': {'term': 'Surgical Margin Resection Status',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2608702,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2608702&version=1.0'}},\n",
       "  'RIN': {'description': 'A numerical assessment of the integrity of RNA based on the entire electrophoretic trace of the RNA sample including the presence or absence of degradation products.\\n',\n",
       "   'termDef': {'term': 'Biospecimen RNA Integrity Number Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5278775,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5278775&version=1.0'}},\n",
       "  'sample_type': {'description': 'Text term to describe the source of a biospecimen used for a laboratory test.\\n',\n",
       "   'termDef': {'term': 'Specimen Type Collection Biospecimen Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3111302,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3111302&version=2.0'}},\n",
       "  'sample_type_id': {'description': 'The accompanying sample type id for the sample type.\\n'},\n",
       "  'section_location': {'description': 'Tissue source of the slide.\\n'},\n",
       "  'sequencing_center': {'description': 'Name of the center that provided the sequence files.\\n'},\n",
       "  'shortest_dimension': {'description': 'Numeric value that represents the shortest dimension of the sample, measured in millimeters.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Short Dimension Millimeter Measurement',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432603,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432603&version=1.0'}},\n",
       "  'site_of_resection_or_biopsy': {'description': 'The third edition of the International Classification of Diseases for Oncology, published in 2000, used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms. The description of an anatomical region or of a body part. Named locations of, or within, the body. A system of numbered categories for representation of data.\\n',\n",
       "   'termDef': {'term': 'International Classification of Diseases for Oncology, Third Edition ICD-O-3 Site Code',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3226281,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3226281&version=1.0'}},\n",
       "  'size_selection_range': {'description': 'Range of size selection.\\n'},\n",
       "  'smoking_history': {'description': 'Category describing current smoking status and smoking history as self-reported by a patient.\\n',\n",
       "   'termDef': {'term': 'Smoking History',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2181650,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2181650&version=1.0'}},\n",
       "  'smoking_intensity': {'description': 'Numeric computed value to represent lifetime tobacco exposure defined as number of cigarettes smoked per day x number of years smoked divided by 20\\n',\n",
       "   'termDef': {'term': 'Person Cigarette Smoking History Pack Year Value',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2955385,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2955385&version=1.0'}},\n",
       "  'source_center': {'description': 'Name of the center that provided the item.\\n'},\n",
       "  'spectrophotometer_method': {'description': 'Name of the method used to determine the concentration of purified nucleic acid within a solution.\\n',\n",
       "   'termDef': {'term': 'Purification Nucleic Acid Solution Concentration Determination Method Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3008378,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3008378&version=1.0'}},\n",
       "  'spike_ins_fasta': {'description': 'Name of the FASTA file that contains the spike-in sequences.\\n'},\n",
       "  'spike_ins_concentration': {'description': 'Spike in concentration.\\n'},\n",
       "  'state': {'description': 'The current state of the object.\\n'},\n",
       "  'target_capture_kit_name': {'description': 'Name of Target Capture Kit.\\n'},\n",
       "  'target_capture_kit_vendor': {'description': 'Vendor of Target Capture Kit.\\n'},\n",
       "  'target_capture_kit_catalog_number': {'description': 'Catalog of Target Capture Kit.\\n'},\n",
       "  'target_capture_kit_version': {'description': 'Version of Target Capture Kit.\\n'},\n",
       "  'target_capture_kit_target_region': {'description': 'Target Capture Kit BED file.\\n'},\n",
       "  'therapeutic_agents': {'description': 'Text identification of the individual agent(s) used as part of a prior treatment regimen.\\n',\n",
       "   'termDef': {'term': 'Prior Therapy Regimen Text',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2975232,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2975232&version=1.0'}},\n",
       "  'time_between_clamping_and_freezing': {'description': 'Numeric representation of the elapsed time between the surgical clamping of blood supply and freezing of the sample, measured in minutes.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Clamping and Freezing Elapsed Minute Time',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432611,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432611&version=1.0'}},\n",
       "  'time_between_excision_and_freezing': {'description': 'Numeric representation of the elapsed time between the excision and freezing of the sample, measured in minutes.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Excision and Freezing Elapsed Minute Time',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432612,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432612&version=1.0'}},\n",
       "  'tissue_or_organ_of_origin': {'description': 'Text term that describes the anatomic site of the tumor or disease.\\n',\n",
       "   'termDef': {'term': 'Tumor Disease Anatomic Site',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3427536,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3427536&version=3.0'}},\n",
       "  'tissue_type': {'description': 'Text term that represents a description of the kind of tissue collected with respect to disease status or proximity to tumor tissue.\\n',\n",
       "   'termDef': {'term': 'Tissue Sample Description Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432687,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432687&version=1.0'}},\n",
       "  'to_trim_adapter_sequence': {'description': 'Does the user suggest adapter trimming?\\n'},\n",
       "  'tobacco_smoking_onset_year': {'description': 'The year in which the participant began smoking.\\n',\n",
       "   'termDef': {'term': 'Started Smoking Year',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2228604,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2228604&version=1.0'}},\n",
       "  'tobacco_smoking_quit_year': {'description': 'The year in which the participant quit smoking.\\n',\n",
       "   'termDef': {'term': 'Stopped Smoking Year',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2228610,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2228610&version=1.0'}},\n",
       "  'tobacco_smoking_status': {'description': 'Category describing current smoking status and smoking history as self-reported by a patient.\\n',\n",
       "   'termDef': {'term': 'Patient Smoking History Category',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2181650,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2181650&version=1.0'}},\n",
       "  'total_sequences': {'description': 'A count of the total number of sequences processed.\\n',\n",
       "   'termDef': {'term': 'Total Sequences',\n",
       "    'source': 'FastQC',\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': 'http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/1%20Basic%20Statistics.html'}},\n",
       "  'treatment_anatomic_site': {'description': 'The anatomic site or field targeted by a treatment regimen or single agent therapy.\\n',\n",
       "   'termDef': {'term': 'Treatment Anatomic Site',\n",
       "    'source': None,\n",
       "    'cde_id': None,\n",
       "    'cde_version': None,\n",
       "    'term_url': None}},\n",
       "  'treatment_intent_type': {'description': 'Text term to identify the reason for the administration of a treatment regimen. [Manually-curated]\\n',\n",
       "   'termDef': {'term': 'Treatment Regimen Intent Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2793511,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2793511&version=1.0'}},\n",
       "  'treatment_or_therapy': {'description': 'A yes/no/unknown/not applicable indicator related to the administration of therapeutic agents received before the body specimen was collected.\\n',\n",
       "   'termDef': {'term': 'Therapeutic Procedure Prior Specimen Collection Administered Yes No Unknown Not Applicable Indicator',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 4231463,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=4231463&version=1.0'}},\n",
       "  'treatment_outcome': {'description': 'Text term that describes the patients final outcome after the treatment was administered.\\n',\n",
       "   'termDef': {'term': 'Treatment Outcome Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5102383,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102383&version=1.0'}},\n",
       "  'treatment_type': {'description': 'Text term that describes the kind of treatment administered.\\n',\n",
       "   'termDef': {'term': 'Treatment Method Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5102381,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5102381&version=1.0'}},\n",
       "  'tumor_grade': {'description': 'Numeric value to express the degree of abnormality of cancer cells, a measure of differentiation and aggressiveness.\\n',\n",
       "   'termDef': {'term': 'Neoplasm Histologic Grade',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2785839,\n",
       "    'cde_version': 2,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2785839&version=2.0'}},\n",
       "  'tumor_code': {'description': 'Diagnostic tumor code of the tissue sample source.\\n'},\n",
       "  'tumor_code_id': {'description': 'BCR-defined id code for the tumor sample.\\n'},\n",
       "  'tumor_descriptor': {'description': 'Text that describes the kind of disease present in the tumor specimen as related to a specific timepoint.\\n',\n",
       "   'termDef': {'term': 'Tumor Tissue Disease Description Type',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3288124,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3288124&version=1.0'}},\n",
       "  'tumor_stage': {'description': 'The extent of a cancer in the body. Staging is usually based on the size of the tumor, whether lymph nodes contain cancer, and whether the cancer has spread from the original site to other parts of the body. The accepted values for tumor_stage depend on the tumor site, type, and accepted staging system. These items should accompany the tumor_stage value as associated metadata.\\n',\n",
       "   'termDef': {'term': 'Tumor Stage',\n",
       "    'source': 'NCIt',\n",
       "    'cde_id': 'C16899',\n",
       "    'cde_version': None,\n",
       "    'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/pages/concept_details.jsf?dictionary=NCI%20Thesaurus&code=C16899'}},\n",
       "  'UUID': {'description': 'A 128-bit identifier. Depending on the mechanism used to generate it, it is either guaranteed to be different from all other UUIDs/GUIDs generated until 3400 AD or extremely likely to be different. Its relatively small size lends itself well to sorting, ordering, and hashing of all sorts, storing in databases, simple allocation, and ease of programming in general.\\n',\n",
       "   'termDef': {'term': 'Universally Unique Identifier',\n",
       "    'source': 'NCIt',\n",
       "    'cde_id': 'C54100',\n",
       "    'cde_version': None,\n",
       "    'term_url': 'https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&version=16.02d&ns=NCI_Thesaurus&code=C54100'}},\n",
       "  'vascular_invasion_present': {'description': 'The yes/no indicator to ask if large vessel or venous invasion was detected by surgery or presence in a tumor specimen.\\n',\n",
       "   'termDef': {'term': 'Tumor Vascular Invasion Ind-3',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 64358,\n",
       "    'cde_version': 3,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=64358&version=3.0'}},\n",
       "  'vital_status': {'description': 'The survival state of the person registered on the protocol.\\n',\n",
       "   'termDef': {'term': 'Patient Vital Status',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5,\n",
       "    'cde_version': 5,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5&version=5.0'}},\n",
       "  'weight': {'description': 'The weight of the patient measured in lbs (HARMONIZED).\\n',\n",
       "   'termDef': {'term': 'Patient Weight Measurement',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 651,\n",
       "    'cde_version': 4,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=651&version=4.0'}},\n",
       "  'well_number': {'description': 'Numeric value that represents the the well location within a plate for the analyte or aliquot from the sample.\\n',\n",
       "   'termDef': {'term': 'Biospecimen Analyte or Aliquot Plate Well Number',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 5432613,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=5432613&version=1.0'}},\n",
       "  'workflow_type': {'description': 'Generic name for the workflow used to analyze a data set.\\n'},\n",
       "  'year_of_birth': {'description': 'Numeric value to represent the calendar year in which an individual was born.\\n',\n",
       "   'termDef': {'term': 'Year Birth Date Number',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2896954,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2896954&version=1.0'}},\n",
       "  'year_of_diagnosis': {'description': \"Numeric value to represent the year of an individual's initial pathologic diagnosis of cancer.\\n\",\n",
       "   'termDef': {'term': 'Year of initial pathologic diagnosis',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2896960,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2896960&version=1.0'}},\n",
       "  'year_of_death': {'description': 'Numeric value to represent the year of the death of an individual.\\n',\n",
       "   'termDef': {'term': 'Year Death Number',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 2897030,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=2897030&version=1.0'}},\n",
       "  'years_smoked_gt89': {'description': 'Indicate whether the numeric value to represent the number of years a person has been smoking (HARMONIZED) is greater than 89 years.\\n',\n",
       "   'termDef': {'term': 'Person Smoking Duration Year Count',\n",
       "    'source': 'caDSR',\n",
       "    'cde_id': 3137957,\n",
       "    'cde_version': 1,\n",
       "    'term_url': 'https://cdebrowser.nci.nih.gov/CDEBrowser/search?elementDetails=9&FirstTimer=0&PageId=ElementDetailsGroup&publicId=3137957&version=1.0'}}},\n",
       " 'lipidomics_file.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'lipidomics_file',\n",
       "  'title': 'Lipidomics File',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'data_file',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Data file containing lipidomics data',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'exclusive': False,\n",
       "    'required': True,\n",
       "    'subgroup': [{'name': 'samples',\n",
       "      'backref': 'lipidomics_files',\n",
       "      'label': 'data_from',\n",
       "      'target_type': 'sample',\n",
       "      'multiplicity': 'many_to_many',\n",
       "      'required': False},\n",
       "     {'name': 'core_metadata_collection',\n",
       "      'backref': 'lipidomics_files',\n",
       "      'label': 'describes',\n",
       "      'target_type': 'core_metadata_collection',\n",
       "      'multiplicity': 'one_to_one',\n",
       "      'required': True}]}],\n",
       "  'required': ['type',\n",
       "   'submitter_id',\n",
       "   'core_metadata_collection',\n",
       "   'data_type',\n",
       "   'data_format',\n",
       "   'data_category'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'$ref': '_definitions.yaml#/data_file_properties',\n",
       "   'samples': {'$ref': '_definitions.yaml#/to_many'},\n",
       "   'core_metadata_collection': {'$ref': '_definitions.yaml#/to_one'},\n",
       "   'data_type': {'description': 'The type of data in this data file',\n",
       "    'enum': ['TIC/XIC', 'MS', 'MS/MS', 'Image']},\n",
       "   'data_format': {'description': 'The format of the data in this data file',\n",
       "    'enum': ['csv', 'tsv', 'xls', 'xlsx']},\n",
       "   'data_category': {'description': 'Broad categorization of the contents of the data file.',\n",
       "    'enum': ['mass spec raw', 'mass spec analysed', 'summarised results']},\n",
       "   'cv': {'description': 'Coefficient of variation (%)', 'type': 'number'}}},\n",
       " 'publication.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'publication',\n",
       "  'title': 'Publication',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Publication for a project',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'publications',\n",
       "    'label': 'refers_to',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_many',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_many_project'}}},\n",
       " 'subject.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'subject',\n",
       "  'title': 'Subject',\n",
       "  'type': 'object',\n",
       "  'namespace': 'http://commons.heartdata.baker.edu.au/',\n",
       "  'category': 'clinical',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'An individual participant in the study with baseline measurements.',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'subjects',\n",
       "    'label': 'part_of',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'required': ['type', 'submitter_id', 'projects', 'baseline_year'],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_one_project'},\n",
       "   'consent_codes': {'description': 'Data Use Restrictions that are used to indicate  permissions/restrictions for datasets and/or materials, and relates to the purposes for which datasets and/or material might be removed, stored or used. \\\\n\\n        Based on the Data Use Ontology : see http://www.obofoundry.org/ontology/duo.html',\n",
       "    'type': 'array'},\n",
       "   'baseline_year': {'description': 'year when baseline measurements were taken',\n",
       "    'type': 'string'},\n",
       "   'cohort_id': {'description': 'Identifier for the cohort the subject is a part of',\n",
       "    'type': 'string'}}},\n",
       " 'core_metadata_collection.yaml': {'$schema': 'http://json-schema.org/draft-04/schema#',\n",
       "  'id': 'core_metadata_collection',\n",
       "  'title': 'Core Metadata Collection',\n",
       "  'type': 'object',\n",
       "  'namespace': 'https://dcp.bionimbus.org/',\n",
       "  'category': 'administrative',\n",
       "  'program': '*',\n",
       "  'project': '*',\n",
       "  'description': 'Structured description of a collection of several dataset\\n',\n",
       "  'additionalProperties': False,\n",
       "  'submittable': True,\n",
       "  'validators': None,\n",
       "  'systemProperties': ['id',\n",
       "   'project_id',\n",
       "   'state',\n",
       "   'created_datetime',\n",
       "   'updated_datetime'],\n",
       "  'links': [{'name': 'projects',\n",
       "    'backref': 'core_metadata_collections',\n",
       "    'label': 'data_from',\n",
       "    'target_type': 'project',\n",
       "    'multiplicity': 'many_to_one',\n",
       "    'required': True}],\n",
       "  'uniqueKeys': [['id'], ['project_id', 'submitter_id']],\n",
       "  'required': ['submitter_id', 'type', 'projects'],\n",
       "  'properties': {'$ref': '_definitions.yaml#/ubiquitous_properties',\n",
       "   'contributor': {'description': 'An node responsible for making contributions to the resource. Examples of a Contributor include a person, an organization, or a service. Typically, the name of a Contributor should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'coverage': {'description': 'The spatial or temporal topic of the resource, the spatial applicability of the resource, or the jurisdiction under which the resource is relevant. Spatial topic and spatial applicability may be a named place or a location specified by its geographic coordinates. Temporal topic may be a named period, date, or date range. A jurisdiction may be a named administrative node or a geographic place to which the resource applies. Recommended best practice is to use a controlled vocabulary such as the Thesaurus of Geographic Names [TGN] (http://www.getty.edu/research/tools/vocabulary/tgn/index.html). Where appropriate, named places or time periods can be used in preference to numeric identifiers such as sets of coordinates or date ranges.\\n',\n",
       "    'type': 'string'},\n",
       "   'creator': {'description': 'An node primarily responsible for making the resource. Examples of a Creator include a person, an organization, or a service. Typically, the name of a Creator should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'date': {'$ref': '_definitions.yaml#/datetime'},\n",
       "   'description': {'description': 'An account of the resource. Description may include but is not limited to: an abstract, a table of contents, a graphical representation, or a free-text account of the resource.\\n',\n",
       "    'type': 'string'},\n",
       "   'format': {'description': 'The file format, physical medium, or dimensions of the resource. Examples of dimensions include size and duration. Recommended best practice is to use a controlled vocabulary such as the list of Internet Media Types [MIME] (http://www.iana.org/assignments/media-types/).\\n',\n",
       "    'type': 'string'},\n",
       "   'language': {'description': 'A language of the resource. Recommended best practice is to use a controlled vocabulary such as RFC 4646 (http://www.ietf.org/rfc/rfc4646.txt).\\n',\n",
       "    'type': 'string'},\n",
       "   'publisher': {'description': 'An node responsible for making the resource available. Examples of a Publisher include a person, an organization, or a service. Typically, the name of a Publisher should be used to indicate the node.\\n',\n",
       "    'type': 'string'},\n",
       "   'relation': {'description': 'A related resource. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\xa0\\n',\n",
       "    'type': 'string'},\n",
       "   'rights': {'description': 'Information about rights held in and over the resource. Typically, rights information includes a statement about various property rights associated with the resource, including intellectual property rights.\\n',\n",
       "    'type': 'string'},\n",
       "   'source': {'description': 'A related resource from which the described resource is derived. The described resource may be derived from the related resource in whole or in part. Recommended best practice is to identify the related resource by means of a string conforming to a formal identification system.\\n',\n",
       "    'type': 'string'},\n",
       "   'subject': {'description': 'The topic of the resource. Typically, the subject will be represented using keywords, key phrases, or classification codes. Recommended best practice is to use a controlled vocabulary.\\n',\n",
       "    'type': 'string'},\n",
       "   'title': {'description': 'A name given to the resource. Typically, a Title will be a name by which the resource is formally known.\\n',\n",
       "    'type': 'string'},\n",
       "   'data_type': {'description': 'The nature or genre of the resource. Recommended best practice is to use a controlled vocabulary such as the DCMI Type Vocabulary [DCMITYPE]. To describe the file format, physical medium, or dimensions of the resource, use the Format element.\\n',\n",
       "    'type': 'string'},\n",
       "   'projects': {'$ref': '_definitions.yaml#/to_one_project'}}}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "write_yaml()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
